---
title: "Univarite MR of traits to MD"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(data.table)
library(vroom)

library(tidyr)
library(purrr)
library(tibble)
library(dplyr)

library(TwoSampleMR)
library(ggplot2)
library(cowplot)
library(wesanderson)
```

```{r}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("set_paths.R")
set_paths(currently_working_env)

# metafile
data_lookup<-read_csv(paste0("metadata/data_lookup.csv"), show_col_types = FALSE) 

# load functions
source("functions.R")
source("functions_mvmr.R")

```



# Run MR for all in a loop 

```{r}

# specify group to process if data is in textfiles
current_trait_category <- "md_unadj"
outcomes <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(!is.na(full_data)) %>%  pull(trait_file_name) 

md_instruments <- data_lookup %>% filter(trait_category == current_trait_category) %>% pull(tophits_data)
md_snps<- bind_rows(lapply(paste0(data_path_tophits, md_instruments), read_tsv))

```



```{r}

traits <- data_lookup %>%  filter(trait_category %in% c("hormones", "reproductive_traits")) %>% pull(trait_file_name)

 ### direction 1 trait - MD

direction1 <-tibble()
sens1 <-tibble()
for (trait_name in traits){
  
  
   dat <- data_lookup %>% filter(trait_file_name == trait_name) 
   
   if (is.na(dat$mrbase.id)){ # text file data
      trait_as_exp = dat %>% pull(tophits_data)
      trait_as_exp <- read_tsv(paste0(data_path_tophits, trait_as_exp),  show_col_types = F)
      
  
     } else { # mrbase data
       
      data_id = dat %>% pull(mrbase.id)
      trait_as_exp <- extract_instruments(data_id) %>% clump_data()
      
     }
     
   
   trait_ss <- dat %>% pull(sample_size)


  for (out in outcomes){
    
    print(paste0("=========Doing ", trait_name, " against ", out))
  
    # read exposure and outcome
    
    exposure_dat <- trait_as_exp
    
    md_pheno_as_outcome <- data_lookup %>% 
          filter(trait_file_name == out) %>%  filter(trait_category == current_trait_category) %>% pull(full_data)
    outcome <- vroom(paste0(data_path_gwas, md_pheno_as_outcome), show_col_types = F)
      
    outcome_dat <- outcome %>% filter(SNP %in% exposure_dat$SNP)
    
    # harmonise
    
    harmonised <- harmonise_data(exposure_dat, outcome_dat )
    
    # do steiger
      
    streiger =  calc_steiger(harmonised, exposure_ss = trait_ss, outcome_ss = 24192) 
    #streiger$summary
    
    # do mr
    res <- mr(harmonised, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) %>% 
           split_outcome() %>% 
           split_exposure() %>% 
           generate_odds_ratios()
    
    direction1<- bind_rows(direction1, res)
    
    
    # do sensitiviry
    out_sensitivity<-
      full_join(mr_pleiotropy_test(harmonised),
                mr_heterogeneity(harmonised, method_list=c("mr_egger_regression", "mr_ivw"))) %>% 
      select(-c("id.outcome", "id.exposure")) %>% filter(method=="Inverse variance weighted") %>% 
      left_join(streiger$summary)
  
    sens1 <- bind_rows(sens1, out_sensitivity )
  
  }
   }




### direction 2 MD-trait

direction2 <-tibble()
sens2 <-tibble()

for (out in outcomes){

  
  for (trait_name in traits){
    
     print(paste0("=========Doing ", out, " against ", trait_name))

    
     dat <- data_lookup %>% filter(trait_file_name == trait_name) 
     
     if (is.na(dat$mrbase.id)){ # text file data

      trait_as_out = dat %>% pull(full_data)
      trait_as_out <- vroom(paste0(data_path_gwas, trait_as_out),  show_col_types = F)
  
     } else { # mrbase data
       
      data_id = dat %>% pull(mrbase.id)
      trait_as_out <- extract_outcome_data(snps = unique(md_snps$SNP), 
                                              outcomes = data_id)
     }
     
     trait_ss <- dat %>% pull(sample_size)
      
    
    md_pheno_as_exposure <- data_lookup %>% 
          filter(trait_file_name == out) %>%  filter(trait_category == current_trait_category) %>% pull(tophits_data)
        
    exposure_dat <- read_tsv(paste0(data_path_tophits, md_pheno_as_exposure),  show_col_types = F)
    
    outcome <- trait_as_out
    outcome_dat <- outcome %>% filter(SNP %in% exposure_dat$SNP)
    
    # harmonise
    
    harmonised <- harmonise_data(exposure_dat, outcome_dat )
    
    # do steiger
      
    streiger =  calc_steiger(harmonised, exposure_ss = 24192, outcome_ss = trait_ss)
    streiger$summary
    
    # do mr
    res <- mr(harmonised, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) %>% 
           split_outcome() %>% 
           split_exposure() %>% 
           generate_odds_ratios()
    
    direction2<- bind_rows(direction2, res)
    
    
    # do sensitiviry
    out_sensitivity<-
      full_join(mr_pleiotropy_test(harmonised),
                mr_heterogeneity(harmonised, method_list=c("mr_egger_regression", "mr_ivw"))) %>% 
      select(-c("id.outcome", "id.exposure"))   %>% filter(method=="Inverse variance weighted") %>% 
      left_join(streiger$summary)    
    
    sens2 <- bind_rows(sens2, out_sensitivity )
  
  }

}  

direction1$dir <- 1
direction2$dir <- 2
sens1$dir <- 1
sens2$dir <- 2

mr_results<- bind_rows(direction1, direction2) %>% distinct()
sens_results<- bind_rows(sens1, sens2)

write_tsv(mr_results, paste0(results_path, current_trait_category,"/", current_trait_category,"_w_traits_biMR_results.tsv"))
write_tsv(sens_results, paste0(results_path, current_trait_category,"/", current_trait_category,"_w_traits_biMR_results_sens.tsv"))
```


```{r}

dat<- read_tsv( paste0(results_path, current_trait_category,"/", current_trait_category,"_w_traits_biMR_results.tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      filter(dir == 1) %>% 
      mutate(exposure = ifelse(exposure == "Age when periods started (menarche)", "Age at menarche (UKB)", exposure)) %>% 
      mutate(outcome = ifelse(outcome == "Age when periods started (menarche)", "Age at menarche (UKB)", outcome)) %>%
      filter(exposure!="Age at menarche") %>% 
      mutate(beta_CI = paste0(round(b,2), " [",round(lo_ci,2) ,":",round(up_ci,2), "]")) %>% 
      mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                                ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 

pal<-rev(c(unname(yarrr::piratepal("pony"))))[c(2,1)]
p<-ggplot(dat, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  #xlim(-1.2,0.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3) +
  geom_text(aes(label=beta_CI),hjust=-0.1, vjust=-0.6, size =3, color = '#333232')+
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  scale_y_discrete(position = "left")+
  labs(color = "",y = "", x = "Beta",
       title= paste0("Univariate MR, 95% CI; direction 1: trait -> MD") )+
       #title= paste0("Univariate MR, 95% CI; direction 2: MD -> trait") )+
  theme(legend.position = "none", plot.title.position  = "plot")

#ggsave(paste0("figures/BMI-to-",current_trait_category ,"_IVW.png"),
#       plot=p, scale=2, 
#       width=10, height=5,
#       units=c("cm"), dpi=200, limitsize=F)
```



# MVMR cBMI + trait -> MD
```{r}

early_bmi_file <- data_lookup %>% filter(trait == "Childhood body size") %>% pull(tophits_data)
early_bmi_exp <- read_tsv(paste0(data_path_tophits, early_bmi_file))
early_bmi_gwas_file <- data_lookup %>% filter(trait == "Childhood body size") %>% pull(full_data)
early_bmi_gwas <- vroom(paste0(data_path_gwas, early_bmi_gwas_file))

traits <- data_lookup %>%  filter(trait_category %in% c("hormones", "reproductive_traits")) %>% pull(trait_file_name)
traits<-traits[2]




mvmr <-tibble()
mvmr_sens <-tibble()

for (trait_name in traits){
  
  # get trait SNPs
   dat <- data_lookup %>% filter(trait_file_name == trait_name) 
   
    if (is.na(dat$mrbase.id)){ # text file data
      trait_as_exp = dat %>% pull(tophits_data)
      trait_as_exp <- read_tsv(paste0(data_path_tophits, trait_as_exp),  show_col_types = F)
      trait_as_out =  dat %>% pull(full_data)
      trait_as_out <- vroom(paste0(data_path_gwas, trait_as_out))

      
     } else { # mrbase data
      data_id = dat %>% pull(mrbase.id)
      trait_as_exp <- extract_instruments(data_id) %>% clump_data()
     }


  for (out in outcomes){
    
    print(paste0("=========Doing MVMR of ", trait_name, "  and childhood BMI against ", out))
  
    # read  outcome
    exposure_dat <- trait_as_exp
    md_pheno_as_outcome <- data_lookup %>% 
          filter(trait_file_name == out) %>%  
          filter(trait_category == current_trait_category) %>% pull(full_data)
    outcome <- vroom(paste0(data_path_gwas, md_pheno_as_outcome), show_col_types = F)
      
    # put all exposure and full gwas dat into lists
      exposure_list <- list(early_bmi_exp, trait_as_exp)
      
      if (!is.na(dat$mrbase.id)){
        trait_as_out <- extract_outcome_data(snps = exposure_list %>% 
                                                purrr::reduce(bind_rows) %>% pull(SNP), 
                                         outcomes = data_id)
        trait_as_out$outcome <- trait_name
      }
      
      full_gwas_list <- list(early_bmi_gwas, trait_as_out)
    
    
    # create exposure_dat format
      exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T) 
    
      #Next, also extract those SNPs from the outcome.
      outcome_dat <- outcome %>%  filter(SNP %in% exposure_dat$SNP)
      
      
      #Once the data has been obtained, harmonise so that all are on the same reference allele.
      mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)
      
      #Finally, perform the multivariable MR analysis
      res <- mv_multiple(mvdat)
      
      mv_res<- res$result %>%
                    split_outcome() %>% 
                    separate(outcome, "outcome", sep="[(]") %>% 
                    generate_odds_ratios() %>% 
                    select(-id.exposure, -id.outcome) %>% 
                    mutate(pair = trait_name)
      
      
      ### sensitivity analysis
      
      # create MVMR package input
      mvmr_input <- make_mvmr_input(exposure_dat,  outcome.data = outcome_dat)
  
      # format data to be in MVMR package-compatible df
      mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta")),  # exposure betas
                              BYG = mvmr_input$YG$beta.outcome,                     # outcome beta
                              seBXGs = mvmr_input$XGs %>% select(contains("se")),  # exposure SEs
                              seBYG = mvmr_input$YG$se.outcome,                     # outcome SEs
                              RSID = mvmr_input$XGs$SNP)                            # SNPs

      print("No phenotypic correlation available, going to use gencov=0 ")
      gen_cov <- 0
      
      #Test for weak instruments
      sres <- strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
      colnames(sres) = paste(mvmr_input$exposures, "(Fst)")
      print(sres)
      
      #Test for horizontal pleiotropy
      pres <- pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)
      
      mvmr_sens_df <- sres
      mvmr_sens_df$Qstat <- pres$Qstat
      mvmr_sens_df$Qpval <- pres$Qpval
      

    mvmr<- bind_rows(mvmr, mv_res)
    mvmr_sens <- bind_rows(mvmr_sens, mvmr_sens_df)
  
  }
}

#write_tsv(mvmr, paste0(results_path,"md_w_traits/","mvmr_results_w_cBMI.tsv")) # adjusted gwas
write_tsv(mvmr, paste0(results_path,"md_w_traits/","mvmr_results_w_cBMI_unadjMD.tsv"))
write_tsv(mvmr_sens_df, paste0(results_path,"md_w_traits/","mvmr_sens_results_w_cBMI_unadjMD.tsv"))


```

```{r}
dat<- read_tsv(paste0(results_path, "md_w_traits/","mvmr_results_w_cBMI_unadjMD.tsv")) %>% 
      mutate(exposure = ifelse(exposure == "age_at_menarche_perry", "Age at menarche (Perry)", exposure)) %>% 
      mutate(exposure = ifelse(exposure == "age_at_menarche_ukb", "Age at menarche (UKB)", exposure)) %>%
  filter(exposure != "Childhood body size") %>% 
  
      mutate(beta_CI = paste0(round(b,2), " [",round(lo_ci,2) ,":",round(up_ci,2), "]")) %>% 
      mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                                ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 

pal<-rev(c(unname(yarrr::piratepal("pony"))))[c(2,1)]
p<-ggplot(dat, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  #xlim(-1.2,0.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3) +
  geom_text(aes(label=beta_CI),hjust=-0.1, vjust=-0.6, size =3, color = '#333232')+
  theme_minimal_hgrid(10, rel_small = 1) +
  facet_wrap(~outcome, ncol=1)+
  scale_y_discrete(position = "left")+
  labs(color = "",y = "", x = "Beta",
       #title= paste0("Univariate MR, 95% CI; direction 1: trait -> MD") )+
       title= paste0("MVMR direct effect, accounted for Childhood body size, 95% CI") )+
  theme(legend.position = "none", plot.title.position  = "plot")
```


# combined plot of total and direcrt effect of childhood body size and age at menarche

```{r}
current_trait_category <- "md_unadj"
# cBMI -> MD total effect
dat_total1<- read_tsv(paste0(results_path, current_trait_category, "/merged_mr_BMI-to-", current_trait_category,".tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      filter(exposure == "Childhood body size") %>% 
      mutate(effect = "Total effect")

# ageatM -> MD total effect
dat_total2<- read_tsv( paste0(results_path, current_trait_category,"/", current_trait_category,"_w_traits_biMR_results.tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      filter(dir == 1) %>% 
      filter(exposure == "Age when periods started (menarche)") %>% # ukb
      mutate(exposure = ifelse(exposure == "Age when periods started (menarche)", "Age at menarche", exposure)) %>%
      mutate(effect = "Total effect")


# mvmr results
dat_direct<- read_tsv(paste0(results_path, "md_w_traits/","mvmr_results_w_cBMI_unadjMD.tsv")) %>% 
      mutate(exposure = ifelse(exposure == "age_at_menarche_ukb", "Age at menarche", exposure)) %>%
      filter(pair == "age_at_menarche_ukb") %>% 
      mutate(effect = "Direct effect") 


dat_merged <- bind_rows(dat_total1, dat_total2, dat_direct) %>%
              select(effect,  everything(), -method) %>% 
              mutate(effect=factor(effect, levels = c("Total effect", "Direct effect"))) %>% 
              mutate(exposure=factor(exposure, levels = c("Childhood body size",  "Age at menarche")),
                     exposure = factor(exposure, levels = rev(levels(exposure)))) %>% 
              arrange(exposure) %>% 
              mutate(beta_CI = paste0(round(b,2), " [",round(lo_ci,2) ,":",round(up_ci,2), "]")) %>% 
              mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                                        ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 


pal<-c(unname(yarrr::piratepal("pony")))[c(8,2)]
p<-ggplot(dat_merged, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  #xlim(-1.2,0.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3) +
  geom_text(aes(label=beta_CI),hjust=0.5, vjust=-1, size =2.9, color = '#3B3B3B')+
  theme_minimal_hgrid(10, rel_small = 1) +
  panel_border() +
  facet_grid(rows=vars(outcome), cols=vars(effect)) +
  scale_y_discrete(position = "right")+
  labs(color = "Exposure",y = "Outcome", x = "effect size"
       #title= paste0("Univariate MR, 95% CI; direction 1: trait -> MD") )+
       #title= paste0("MVMR direct effect, accounted for Childhood body size, 95% CI"
                      )+
  theme(legend.position = "bottom", legend.justification = "center",
        plot.title.position  = "plot",
        strip.text.x = element_text(face = "bold"),
        strip.text.y = element_text(face = "bold"),
        axis.text.y=element_blank())+
  guides(shape="none", colour=guide_legend(reverse = T))


```



# combined plot of total and direcrt effect of childhood body size and adult body size

```{r}
current_trait_category <- "md_unadj"
# cBMI or aBMI-> MD total effect
dat_total<- read_tsv(paste0(results_path, current_trait_category, "/merged_mr_BMI-to-", current_trait_category,".tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      #filter(exposure == "Childhood body size") %>% 
      mutate(effect = "Total effect")


# mvmr results
dat_direct<- read_tsv(paste0(results_path, current_trait_category, "/mvmr/merged/merged_mvmr_BMIs-", current_trait_category,"-to-BCAC.tsv")) %>% 
  mutate(effect = "Direct effect") 


dat_merged_BMIS <- bind_rows(dat_total, dat_direct) %>%
              select(effect,  everything(), -method) %>% 
              mutate(effect=factor(effect, levels = c("Total effect", "Direct effect"))) %>% 
              mutate(exposure=factor(exposure, levels = c("Childhood body size",  "Adult body size")),
                     exposure = factor(exposure, levels = rev(levels(exposure)))) %>% 
              arrange(exposure) %>% 
              mutate(beta_CI = paste0(round(b,2), " [",round(lo_ci,2) ,":",round(up_ci,2), "]")) %>% 
              mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                                        ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 


pal<-c(unname(yarrr::piratepal("pony")))[c(1,2)]
p<-ggplot(dat_merged_BMIS, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  #xlim(-1.2,0.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3) +
  geom_text(aes(label=beta_CI),hjust=0.5, vjust=-1, size =2.9, color = '#3B3B3B')+
  theme_minimal_hgrid(10, rel_small = 1) +
  panel_border() +
  facet_grid(rows=vars(outcome), cols=vars(effect), switch = "y") +
  scale_y_discrete(position = "left")+
  labs(color = "Exposure",y = "Outcome", x = "effect size"
       #title= paste0("Univariate MR, 95% CI; direction 1: trait -> MD") )+
       #title= paste0("MVMR direct effect, accounted for Childhood body size, 95% CI"
                      )+
  theme(legend.position = "bottom", legend.justification = "center",
        plot.title.position  = "plot",
        strip.text.x = element_text(face = "bold"),
        strip.text.y = element_text(face = "bold"),
        axis.text.y=element_blank())+
  guides(shape="none", colour=guide_legend(reverse = T))


```


# try a combined plot of cBMI, aBMI, aaM - total and direct effects

```{r}

dat_all <- bind_rows(dat_merged %>% mutate(cov = "c.       Direct effect (MVMR) "),
                     dat_merged_BMIS %>% mutate(cov = "b.       Direct effect (MVMR)")) %>% 
  select(-id.exposure, -id.outcome, -dir, -pair, -outcome.full ) %>% 
  mutate(cov = ifelse(effect == "Total effect", "a.        Total effect (MR)", cov)) %>% 
    distinct() %>% 
  mutate(cov = factor(cov, levels = c("a.        Total effect (MR)",
                                      "b.       Direct effect (MVMR)", 
                                      "c.       Direct effect (MVMR) "))) %>% 
  mutate(exposure = factor(exposure, levels = rev(c("Childhood body size", 
                                                "Adult body size",
                                                "Age at menarche"))))

pal<-c(unname(yarrr::piratepal("pony")))[c(1,8,2)]
pX<-ggplot(dat_all, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  xlim(-1.2,1.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3, color = '#3B3B3B') +
  geom_text(aes(label=beta_CI),hjust=0.28, vjust=-1, size =2.5, color = '#636363')+
  theme_minimal_hgrid(9, rel_small = 1) +
  #panel_border() +
  facet_grid(rows=vars(outcome), cols=vars(cov), switch = "y") +
  scale_y_discrete(position = "left")+
  labs(color = "Exposure",y = "Outcome",
       x = "beta coefficient (1 SD change in MD phenotype per body size category or SD age change)")+
  theme(legend.position = "right", legend.justification = "center",
        plot.title.position  = "plot",
        strip.text.x = element_text(face = "bold"),
        strip.text.y = element_text(face = "bold"),
        axis.text.y=element_blank(),
        panel.spacing = unit(1, "lines"))+
  guides(shape="none", colour=guide_legend(reverse = T))
```




# combined plot of total and direcrt effect of childhood body size and IGF

```{r}
current_trait_category <- "md_unadj"
# cBMI -> MD total effect
dat_total1<- read_tsv(paste0(results_path, current_trait_category, "/merged_mr_BMI-to-", current_trait_category,".tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      filter(exposure == "Childhood body size") %>% 
      mutate(effect = "Total effect")

# IGF -> MD total effect
dat_total2<- read_tsv( paste0(results_path, current_trait_category,"/", current_trait_category,"_w_traits_biMR_results.tsv")) %>% 
      filter(method == "Inverse variance weighted") %>% 
      filter(dir == 1) %>% 
      filter(exposure == "IGF") %>% # ukb
      mutate(effect = "Total effect")


# mvmr results
dat_direct<- read_tsv(paste0(results_path, "md_w_traits/","mvmr_results_w_cBMI_unadjMD.tsv")) %>% 
      filter(pair == "igf") %>% 
      mutate(effect = "Direct effect") 


dat_merged <- bind_rows(dat_total1, dat_total2, dat_direct) %>%
              select(effect,  everything(), -method) %>% 
              mutate(effect=factor(effect, levels = c("Total effect", "Direct effect"))) %>% 
              mutate(exposure=factor(exposure, levels = c("Childhood body size",  "IGF")),
                     exposure = factor(exposure, levels = rev(levels(exposure)))) %>% 
              arrange(exposure) %>% 
              mutate(beta_CI = paste0(round(b,2), " [",round(lo_ci,2) ,":",round(up_ci,2), "]")) %>% 
              mutate(effect_direction = ifelse(or_lci95 > 1 & or_uci95 >= 1, 'ok',
                                        ifelse(or_lci95 < 1 & or_uci95 <= 1, 'ok', 'overlaps null'))) 


pal<-c(unname(yarrr::piratepal("pony")))[c(9,2)]
p<-ggplot(dat_merged, aes(y=exposure, x=b, label=outcome, shape=effect_direction, colour=exposure)) +
  geom_errorbarh(aes(xmin=lo_ci, xmax=up_ci), height=0.2) +
  geom_point(size=2.5)+
  #xlim(-1.2,0.5)+
  scale_color_manual(values=pal)+
  scale_shape_manual(values=c(16,1))+
  geom_vline(xintercept=0, linetype=3) +
  geom_text(aes(label=beta_CI),hjust=0.5, vjust=-1, size =2.9, color = '#3B3B3B')+
  theme_minimal_hgrid(10, rel_small = 1) +
  panel_border() +
  facet_grid(rows=vars(outcome), cols=vars(effect)) +
  scale_y_discrete(position = "right")+
  labs(color = "Exposure",y = "Outcome", x = "effect size"
       #title= paste0("Univariate MR, 95% CI; direction 1: trait -> MD") )+
       #title= paste0("MVMR direct effect, accounted for Childhood body size, 95% CI"
                      )+
  theme(legend.position = "bottom", legend.justification = "center",
        plot.title.position  = "plot",
        strip.text.x = element_text(face = "bold"),
        strip.text.y = element_text(face = "bold"),
        axis.text.y=element_blank())+
  guides(shape="none", colour=guide_legend(reverse = T))


```



