---
title: "MVMR analysis of BMI and mediators to BC"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(vroom)

library(tidyr)
library(purrr)
library(tibble)
library(dplyr)

library(TwoSampleMR)
library(MVMR)
```

```{r message=F}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("set_paths.R")
set_paths(currently_working_env)

# metafile
data_lookup<-read_csv(paste0("metadata/data_lookup.csv")) 
bcac_lookup<-read_csv(paste0("metadata/data_lookup_BCAC.csv")) 

# load functions
source("functions_mvmr.R")
source("functions.R")

```

# Load all BMI files
```{r, cache=TRUE, message=F}
# Load BMI exposures
early_bmi_file <- data_lookup %>% filter(trait == "Childhood body size") %>% pull(tophits_data)
early_bmi_exp <- read_tsv(paste0(data_path_tophits, early_bmi_file))
dim(early_bmi_exp) ## 115 in new

adult_bmi_file <- data_lookup %>% filter(trait == "Adult body size") %>% pull(tophits_data)### or this could be Adult BMI unadjust if used for MVMR with unadj
adult_bmi_exp <- read_tsv(paste0(data_path_tophits, adult_bmi_file))
dim(adult_bmi_exp) # 173 in new

# Load BMI outcomes
early_bmi_gwas_file <- data_lookup %>% filter(trait == "Childhood body size") %>% pull(full_data)
early_bmi_gwas <- vroom(paste0(data_path_gwas, early_bmi_gwas_file))
dim(early_bmi_gwas) 

adult_bmi_gwas_file <- data_lookup %>% filter(trait == "Adult body size") %>% pull(full_data)
adult_bmi_gwas <- vroom(paste0(data_path_gwas, adult_bmi_gwas_file))
dim(adult_bmi_gwas) 

```






# Run MVMR for each mediator separately in a loop
# can do by mediators category 

```{r}
# specify group to process if data is in textfiles
current_trait_category <- "md_unadj"
mediators <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(!is.na(full_data))  %>%  pull(full_data)
```

# Massive for-loop to:
## For a selected breast cancer outcome (load)
## Iterate over all mediators in current_trait_category, and:
## Run MVMR 1: Childhood BMI and Adult BMI as exposures, Mediator as outcome
## Run MVMR 2: Childhood BMI and Mediator as exposures, Breast cancer as outcome
## Run MVMR 3: Adult BMI     and Mediator as exposures, Breast cancer as outcome
## Run MVMR 4: Childhood BMI, Adult BMI, and Mediator as exposures, Breast cancer as outcome

```{r message= F}
results_path_sub <-   paste0(results_path, current_trait_category, "/")

run_sens = T

run_analysis_1 = T
run_analysis_2 = F
run_analysis_3 = F
run_analysis_4 = F

for (j in 1:length(bcac_lookup$trait)){

  # first select breast cancer outcome
  if (!is.na(bcac_lookup$mr.base.id[j])){
    outcome_source = 'outcome_mrbase'
    outcome_name = bcac_lookup$mr.base.id[j]
    print(paste0( "Using breast cancer data: ", outcome_name))
    breast_cancer_data <- outcome_name
  } else {
    outcome_source = 'outcome_textfile'
    outcome_name = bcac_lookup$full_data[j]
    print(paste0( "Using breast cancer data: ", outcome_name))
    # read in the outcome file
    breast_cancer_data  <- vroom(paste0(data_path_gwas, outcome_name)) 
  }
  outcome_trait_name = bcac_lookup$trait_file_name[j]
  outcome_trait = bcac_lookup$trait[j]
  
  
  # iterate over each mediators, and do any/all of 4 possible analyses
  for (i in 1:length(mediators)){
    
    if( mediators[i] %in% data_lookup$full_data ){
      # mediaotor is in a text file
      format <-"mediator_textfile"
      mediator_name <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(full_data == mediators[i]) %>% pull(trait)
      mediator_file_name <- data_lookup %>% filter(trait_category == current_trait_category) %>%filter(full_data == mediators[i]) %>% pull(trait_file_name)
      
      # load mediator instruments
      tophits_file <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(full_data == mediators[i]) %>% pull(tophits_data)
      exposure_mediator <- vroom(paste0(data_path_tophits, tophits_file), 
                                 col_types = cols(effect_allele.exposure = col_character())) # to avoid T being read as TRUE
                          
                        
        
      # load full GWAS data (outcomes) and subset to exposure SNPs
      outcome_mediator <- vroom(paste0(data_path_gwas, mediators[i]))
  
  
    } else if ( mediators[i] %in% data_lookup$mrbase.id ){
      # mediaotr is in mrbase
      format <- "mediator_mrbase"
      mediator_name <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(mrbase.id == mediators[i]) %>% pull(trait)
      mediator_file_name <- data_lookup %>% filter(trait_category == current_trait_category) %>% filter(mrbase.id == mediators[i]) %>% pull(trait_file_name)
  
      # load mediator instruments
      exposure_mediator <- extract_instruments(mediators[i])
      if (is.null(exposure_mediator)){ stop(paste0("No instruments found for ", mediators[i]))} # if no SNPs returned    
  
      exposure_mediator <- clump_data(exposure_mediator)
      exposure_mediator$exposure <- mediator_name
    }
    print(paste0("Currently processing ", mediator_name, " from " , format ))
    
    # make sure we have a place to write
    mvmr_dir <- paste0(results_path_sub, "mvmr/")
    mvmr_dir_w_backslash <- gsub(" ", "\\ ", mvmr_dir, fixed=T) # create path vector escaping spaces, otherwise sytem call cant process it
    if(!dir.exists(mvmr_dir)){ system(paste("mkdir -p", mvmr_dir_w_backslash))}
    
    #
    #
    #
    
    
    # Analysis 1. Multivariable MR: Childhood BMI and Adult BMI as exposures, Mediator as outcome
    if (run_analysis_1) {
      print("=========== Running analysis 1 =============")
      ## NB this analysis does not use breast cancer data defined in the outer for loop - so really only need to run it once, set to F
    
      # put all exposure and full gwas dat into lists
      exposure_list <- list(early_bmi_exp, adult_bmi_exp)
      full_gwas_list<- list(early_bmi_gwas, adult_bmi_gwas)
    
      # create exposure_dat format
      exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T)
    
      #Next, also extract those SNPs from the outcome.
      if (format == "mediator_mrbase"){
        outcome_dat <- extract_outcome_data(snps = exposure_dat$SNP, 
                                            outcomes = mediators[i])
      } else if (format == "mediator_textfile") {
        outcome_dat <- outcome_mediator %>% filter(SNP %in% exposure_dat$SNP)
      }
      #Once the data has been obtained, harmonise so that all are on the same reference allele.
      mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)
      
      #Finally, perform the multivariable MR analysis
      res_bmis <- mv_multiple(mvdat)
      
      mv_res_bmis<- res_bmis$result %>%
                    split_outcome() %>%
                    generate_odds_ratios() %>% 
                    select(-id.exposure, -id.outcome)
      mv_res_bmis$outcome.full<-mediator_name
      
      write_tsv(mv_res_bmis, paste0(mvmr_dir, "mvmr_BMIs-", mediator_file_name, "_using_2SMR.tsv"))
      
      
      ## sensitivity analysis
      
      if (run_sens){
        print("Doing sensitivity analysis")
        
        mvmr_input <- make_mvmr_input(exposure_dat, outcome.data = outcome_dat)

        # format data to be in MVMR package-compatiable df
        mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta")),  # exposure betas
                                BYG = mvmr_input$YG$beta.outcome,                        # outcome beta
                                seBXGs = mvmr_input$XGs %>% select(contains("se")),      # exposure SEs
                                seBYG = mvmr_input$YG$se.outcome,                        # outcome SEs
                                RSID = mvmr_input$XGs$SNP)                               # SNPs
     
        ## Sensitivity tests
        # find phenotypic correlation value in table and estimate gencov
        print(paste0("Using phenocor for calculating Fst"))
        pheno_corr_bmi =  0.197457971 # externally calculated value
        pheno_mat = matrix(c(1, pheno_corr_bmi, pheno_corr_bmi, 1), nrow=2, ncol=2)
        colnames(pheno_mat) = rownames(pheno_mat) = c("Childhood BMI", 'Adult BMI')
        print(pheno_mat)
        #extract SE matrix
        se_matrix <- mvmr_out %>% as_tibble() %>% select(contains("sebetaX")) %>% as.data.frame()
        #estimate gencov
        gen_cov <- phenocov_mvmr(Pcov = as.matrix(pheno_mat), seBXGs = se_matrix)
          #Test for weak instruments
        sres <- strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
        colnames(sres) = paste(c("Childhood BMI", 'Adult BMI'), "(Fst)")
        print(sres)
        #Test for horizontal pleiotropy
        pres <- pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)
        mvmr_sens_df <- sres
        mvmr_sens_df$Qstat <- pres$Qstat
        mvmr_sens_df$Qpval <- pres$Qpval
        mvmr_sens_df$trait <- unique(outcome_dat$outcome)
        write_tsv(mvmr_sens_df, paste0(mvmr_dir, "mvmr_sens_BMIs-", mediator_file_name, "_using_2SMR.tsv"))
      }
      
      rm(exposure_list, full_gwas_list, exposure_dat, outcome_dat)
    }  
    #
    #
    #
    
    
    # Analysis 2. Multivariable MR: Childhood BMI and Mediator as exposures, Breast cancer as outcome
    if (run_analysis_2) {
    print("=========== Running analysis 2 =============")
      
      # put all exposure and full gwas dat into lists
      exposure_list <- list(early_bmi_exp, exposure_mediator)
      
      
      if (format == "mediator_mrbase"){
        outcome_mediator <- extract_outcome_data(snps = exposure_list %>% 
                                                purrr::reduce(bind_rows) %>% pull(SNP), 
                                         outcomes = mediators[i])
        outcome_mediator$outcome <- mediator_name
      }
      
      full_gwas_list <- list(early_bmi_gwas, outcome_mediator)
      
      # create exposure_dat format
      exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T) 
    
      #Next, also extract those SNPs from the outcome.
      if (outcome_source == 'outcome_mrbase'){
        outcome_dat <- extract_outcome_data(exposure_dat$SNP,
                                            breast_cancer_data)
      } else if (outcome_source == 'outcome_textfile') {
        outcome_dat <- breast_cancer_data %>%  
                          filter(SNP %in% exposure_dat$SNP)
      }
      
      #Once the data has been obtained, harmonise so that all are on the same reference allele.
      mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)
      
      #Finally, perform the multivariable MR analysis
      res <- mv_multiple(mvdat)
      
      mv_res<- res$result %>%
                    split_outcome() %>% 
                    separate(outcome, "outcome", sep="[(]") %>% 
                    generate_odds_ratios() %>% 
                    select(-id.exposure, -id.outcome) %>% 
                    mutate(outcome = outcome_trait,
                           mediator = mediator_file_name)
        
      write_tsv(mv_res, paste0(mvmr_dir, "mvmr_earlyBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,"_using_2SMR.tsv"))
    
      ## sensitivity analysis
      
      if (run_sens) {
         print("Doing sensitivity analysis")

        # create MVMR package input
        if (outcome_source == 'outcome_mrbase'){
          # specify mrbase id
          mvmr_input <- make_mvmr_input(exposure_dat, outcome.id.mrbase = breast_cancer_data)
  
         } else if (outcome_source == 'outcome_textfile') {
          # specify text file
         mvmr_input <- make_mvmr_input(exposure_dat,  outcome.data = breast_cancer_data)
  
        }
        
        
        # format data to be in MVMR package-compatible df
        mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta")),  # exposure betas
                                BYG = mvmr_input$YG$beta.outcome,                     # outcome beta
                                seBXGs = mvmr_input$XGs %>% select(contains("se")),  # exposure SEs
                                seBYG = mvmr_input$YG$se.outcome,                     # outcome SEs
                                RSID = mvmr_input$XGs$SNP)                            # SNPs
    
      #  estimate causal effects using method in MVMR package
        mvmr_res <-ivw_mvmr(r_input=mvmr_out) %>% 
                   tidy_mvmr_output() %>% 
                   mutate(exposure = mvmr_input$exposures,
                          outcome = outcome_trait_name)
        # not saving mvmr outcput from mvmr package- it is the same as from 2smr
        #write_tsv(mvmr_res, paste0(mvmr_dir, "mvmr_earlyBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,"_using_MVMR.tsv"))
        
      
        
        # find phenotypic correlation value in table and estimate gencov
        if (mediator_name %in% phenocor_matrix$mediator){
          print(paste0("Using phenocor for calculating Fst"))
          phenocor_values <- filter(phenocor_matrix, mediator == mediator_name)
          pheno_mat <- matrix(c(1, phenocor_values$early_bmi,
                            phenocor_values$early_bmi, 1), nrow=2, ncol=2)
          colnames(pheno_mat) = rownames(pheno_mat) = c("Childhood body size", mediator_name)
          print(pheno_mat)
          
          #extract SE matrix ### NB use MVMR_OUT now
          se_matrix <- mvmr_out %>% as_tibble() %>% select(contains("sebetaX")) %>% as.data.frame()
      
          #estimate gencov
          gen_cov <- phenocov_mvmr(Pcov = as.matrix(pheno_mat), seBXGs = se_matrix)
        } else{
          print("No phenotypic correlation available, going to use gencov=0 ")
          gen_cov <- 0
        }
        
        #Test for weak instruments
        sres <- strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
        colnames(sres) = paste(c("Childhood body size", mediator_name), "(Fst)")
        print(sres)
        
        #Test for horizontal pleiotropy
        pres <- pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)
        
        mvmr_sens_df <- sres
        mvmr_sens_df$Qstat <- pres$Qstat
        mvmr_sens_df$Qpval <- pres$Qpval
        
        write_tsv(mvmr_sens_df, paste0(mvmr_dir, "mvmr_sens_earlyBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,".tsv"))
      }  
      rm(exposure_list, full_gwas_list, exposure_dat, outcome_dat)
    }
    #
    #
    #
    # Analysis 3. Multivariable MR: Adult BMI and Mediator as exposures, Breast cancer as outcome
    if (run_analysis_3) {
      print("=========== Running analysis 3 =============")
      
      # put all exposure and full gwas dat into lists
      exposure_list <- list(adult_bmi_exp, exposure_mediator)
      
      
      if (format == "mediator_mrbase"){
        outcome_mediator <- extract_outcome_data(snps = exposure_list %>% 
                                                purrr::reduce(bind_rows) %>% pull(SNP), 
                                         outcomes = mediators[i])
        outcome_mediator$outcome <- mediator_name
      }
      
      full_gwas_list <- list(adult_bmi_gwas, outcome_mediator)
      
      # create exposure_dat format
      exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T) 
    
      #Next, also extract those SNPs from the outcome.
       if (outcome_source == 'outcome_mrbase'){
        outcome_dat <- extract_outcome_data(exposure_dat$SNP,
                                            breast_cancer_data)
      } else if (outcome_source == 'outcome_textfile') {
        outcome_dat <- breast_cancer_data %>%  
                          filter(SNP %in% exposure_dat$SNP)
      }
      
      #Once the data has been obtained, harmonise so that all are on the same reference allele.
      mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)
      
      #Finally, perform the multivariable MR analysis
      res <- mv_multiple(mvdat)
      
      mv_res<- res$result %>%
                    split_outcome() %>% 
                    separate(outcome, "outcome", sep="[(]") %>% 
                    generate_odds_ratios() %>% 
                    select(-id.exposure, -id.outcome)%>% 
                    mutate(outcome = outcome_trait,
                           mediator = mediator_file_name)
      
      write_tsv(mv_res, paste0(mvmr_dir, "mvmr_adultBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,"_using_2SMR.tsv"))
    
      ## sensitivity analysis
      if (run_sens) {
        # create MVMR package input
        if (outcome_source == 'outcome_mrbase'){
          # specify mrbase id
          mvmr_input <- make_mvmr_input(exposure_dat, outcome.id.mrbase = breast_cancer_data)
         } else if (outcome_source == 'outcome_textfile') {
          # specify text file
         mvmr_input <- make_mvmr_input(exposure_dat,  outcome.data = breast_cancer_data)
        }
        
        # format data to be in MVMR package-compatible df
        mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta")),  # exposure betas
                                BYG = mvmr_input$YG$beta.outcome,                     # outcome beta
                                seBXGs = mvmr_input$XGs %>% select(contains("se")),  # exposure SEs
                                seBYG = mvmr_input$YG$se.outcome,                     # outcome SEs
                                RSID = mvmr_input$XGs$SNP)                            # SNPs
    
      #  estimate causal effects using method in MVMR package
        mvmr_res <-ivw_mvmr(r_input=mvmr_out) %>% 
                   tidy_mvmr_output() %>% 
                   mutate(exposure = mvmr_input$exposures,
                          outcome = outcome_trait_name)
        # not saving mvmr output from mvmr package- it is the same as from 2smr
        #write_tsv(mvmr_res, paste0(mvmr_dir, "mvmr_adultBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,"_using_MVMR.tsv"))
        
      
        
        # find phenotypic correlation value in table and estimate gencov
        if (mediator_name %in% phenocor_matrix$mediator){
          print(paste0("Using phenocor for calculating Fst"))
          phenocor_values <- filter(phenocor_matrix, mediator == mediator_name)
          pheno_mat <- matrix(c(1, phenocor_values$adult_bmi,
                            phenocor_values$adult_bmi, 1), nrow=2, ncol=2)
          colnames(pheno_mat) = rownames(pheno_mat) = c("Adult body size", mediator_name)
          print(pheno_mat)
          #extract SE matrix ### NB use MVMR_OUT now
          se_matrix <- mvmr_out %>% as_tibble() %>% select(contains("sebetaX")) %>% as.data.frame()
      
          #estimate gencov
          gen_cov <- phenocov_mvmr(Pcov = as.matrix(pheno_mat), seBXGs = se_matrix)
        } else{
          print("No phenotypic correlation available, going to use gencov=0 ")
          gen_cov <- 0
        }
        
        #Test for weak instruments
        sres <- strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
        colnames(sres) = paste(c("Adult body size", mediator_name), "(Fst)")
        print(sres)
        
        #Test for horizontal pleiotropy
        pres <- pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)
        
        mvmr_sens_df <- sres
        mvmr_sens_df$Qstat <- pres$Qstat
        mvmr_sens_df$Qpval <- pres$Qpval
        
        write_tsv(mvmr_sens_df, paste0(mvmr_dir, "mvmr_sens_adultBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,".tsv"))
      } 
      rm(exposure_list, full_gwas_list, exposure_dat, outcome_dat)
    }
    
    #
    #
    #
    
   # Analysis 4. Multivariable MR: Childhood BMI, Adult BMI, and Mediator as exposures, Breast cancer as outcome
  
    if (run_analysis_4) {
    print("=========== Running analysis 4 =============")
    
      exposure_list <- list(early_bmi_exp, adult_bmi_exp, exposure_mediator)
      if (format == "mediator_mrbase"){
        outcome_mediator <- extract_outcome_data(snps = exposure_list %>% purrr::reduce(bind_rows)  %>% pull(SNP), 
                                         outcomes = mediators[i])
        outcome_mediator$outcome <- mediator_name
      }
      
      full_gwas_list<- list(early_bmi_gwas, adult_bmi_gwas,  outcome_mediator)
      
      # create exposure_dat format
      exposure_dat <- get_mv_exposures(exposure_list, full_gwas_list, clump_exposures = T)
    
      #Next, also extract those SNPs from the outcome.
       if (outcome_source == 'outcome_mrbase'){
        outcome_dat <- extract_outcome_data(exposure_dat$SNP,
                                            breast_cancer_data)
      } else if (outcome_source == 'outcome_textfile') {
       outcome_dat <- breast_cancer_data %>%  
                          filter(SNP %in% exposure_dat$SNP)
      }
      
      #Once the data has been obtained, harmonise so that all are on the same reference allele.
      mvdat <- mv_harmonise_data(exposure_dat, outcome_dat)
      
      #Finally, perform the multivariable MR analysis
      res <- mv_multiple(mvdat)
      
      mv_3exp<- res$result %>%
                    split_outcome() %>% 
                    separate(outcome, "outcome", sep="[(]") %>% 
                    generate_odds_ratios() %>% 
                    select(-id.exposure, -id.outcome)%>% 
                    mutate(outcome = outcome_trait,
                           mediator = mediator_file_name)
      
      write_tsv(mv_3exp, paste0(mvmr_dir, "mvmr_adultBMIearlyBMI-", mediator_file_name,"-to-BCAC_", outcome_trait_name,"_using_2SMR.tsv"))
      
      
      ## sensitivity analysis
      if (run_sens) {
        # create MVMR package input
        if (outcome_source == 'outcome_mrbase'){
          # specify mrbase id
          mvmr_input <- make_mvmr_input(exposure_dat, outcome.id.mrbase = breast_cancer_data)
         } else if (outcome_source == 'outcome_textfile') {
          # specify text file
         mvmr_input <- make_mvmr_input(exposure_dat,  outcome.data = breast_cancer_data)
        }
        
        # format data to be in MVMR package-compatiable df
        mvmr_out <- format_mvmr(BXGs = mvmr_input$XGs %>% select(contains("beta")),  # exposure betas
                                BYG = mvmr_input$YG$beta.outcome,                        # outcome beta
                                seBXGs = mvmr_input$XGs %>% select(contains("se")),      # exposure SEs
                                seBYG = mvmr_input$YG$se.outcome,                        # outcome SEs
                                RSID = mvmr_input$XGs$SNP)                               # SNPs
    
        # find phenotypic correlation value in table and estimate gencov
        if (mediator_name %in% phenocor_matrix$mediator){
          print(paste0("Using phenocor for calculating Fst"))
          
          
          y <- phenocor_matrix %>% filter(mediator %in% c(mediator_name, "Childhood body size", "Adult body size")) %>% 
              select(mediator, adult_bmi, early_bmi) %>% # order matters
              mutate(mediator =ifelse(mediator == mediator_name, 'mediator', mediator)) %>% 
              arrange(mediator) %>%  # always bottom row
              column_to_rownames('mediator') 
            
          pheno_mat <- y %>% mutate(mediator = c(as.vector(as.matrix(y['mediator',])), 1)) %>% 
                             rename("Childhood body size" = "early_bmi", "Adult body size" = "adult_bmi")
    
          print(pheno_mat)
          
          #extract SE matrixm ### NB USE MVMR_OUT now
          se_matrix <- mvmr_out %>% as_tibble() %>% select(contains("sebetaX")) %>% as.data.frame()
      
          #estimate gencov
          gen_cov <- phenocov_mvmr(Pcov = as.matrix(pheno_mat), seBXGs = se_matrix)
        } else{
          gen_cov <- 0
        }
        
          #Test for weak instruments
        sres <- strength_mvmr(r_input=mvmr_out, gencov=gen_cov)
        colnames(sres) = paste(mvmr_input$exposures, "(Fst)")
        print(sres)
        
        #Test for horizontal pleiotropy
        pres <- pleiotropy_mvmr(r_input=mvmr_out, gencov=gen_cov)
        
        mvmr_sens_df <- sres
        mvmr_sens_df$Qstat <- pres$Qstat
        mvmr_sens_df$Qpval <- pres$Qpval
        
        write_tsv(mvmr_sens_df, paste0(mvmr_dir, "mvmr_sens_adultBMI-earlyBMI-", mediator_file_name ,"-to-BCAC_", outcome_trait_name,".tsv"))
      }
      
      rm(exposure_list, outcome_mediator, full_gwas_list, exposure_dat, outcome_dat)
    }
    
    
    
    
    print(paste0("Finished analyses for ", mediators[i]))
  }
}
```


# Merge the MVMR results into one table for each trait category

```{r message=FALSE}
# select trait category
current_trait_category <- "md_unadj"
mvmr_files <- list.files(path = paste0(results_path, current_trait_category, '/mvmr'), full.names = T)
# drop sensitivity files and BMI
mvmr_files <- mvmr_files[!grepl('sens', mvmr_files)]

# make sure this dir exists : 02_Results/sleep_trairs/mvmr/merged
mvmr_merged_dir <- paste0(results_path_sub, "mvmr/merged/")
mvmr_merged_dir_w_backslash <- gsub(" ", "\\ ", mvmr_merged_dir, fixed=T) # create path vector escaping spaces, otherwise system call cant process it
if(!dir.exists(mvmr_merged_dir_w_backslash)){ system(paste("mkdir -p", mvmr_merged_dir_w_backslash))}


# this is a list of MVMR types to cycle through
mvmr_types <- c( "mvmr_BMIs-", "mvmr_earlyBMI-", "mvmr_adultBMI-", "mvmr_adultBMIearlyBMI-")
mvmr_types <- c("mvmr_BMIs-", "mvmr_earlyBMI-")

# for each mvmr type, load all indiv trait files, merge, save
for (mvmr_type in mvmr_types) {

  print(paste0("Current trait category: ", current_trait_category))
  
  # select type
  mvmr_files_type <- mvmr_files[grepl(mvmr_type, mvmr_files )]
  
  
  print(paste0("Reading data from analysis: ", mvmr_type))
  
  # read all individual mediators
  l <- lapply(mvmr_files_type, read_tsv, show_col_types = FALSE)
  df <- l %>% purrr::reduce(bind_rows) 
  
  file_prefix <- paste0(mvmr_type, current_trait_category, "-to-BCAC")
  write_tsv(df, paste0(results_path, current_trait_category, "/mvmr/merged/merged_", file_prefix, ".tsv"))
  print(paste0(" -> Merged and saved."))

}
```

# Merge sens analyses

```{r message=FALSE}
# select trait category
current_trait_category <- "md_unadj"
mvmr_files <- list.files(path = paste0(results_path, current_trait_category, '/mvmr'), full.names = T)
# drop sensitivity files and BMI
mvmr_files <- mvmr_files[grepl('sens', mvmr_files)]

# make sure this dir exists : 02_Results/sleep_trairs/mvmr/merged
mvmr_merged_dir <- paste0(results_path_sub, "mvmr/merged/")
mvmr_merged_dir_w_backslash <- gsub(" ", "\\ ", mvmr_merged_dir, fixed=T) # create path vector escaping spaces, otherwise system call cant process it
if(!dir.exists(mvmr_merged_dir_w_backslash)){ system(paste("mkdir -p", mvmr_merged_dir_w_backslash))}


# this is a list of MVMR types to cycle through
#mvmr_types <- c( "mvmr_BMIs-", "mvmr_earlyBMI-", "mvmr_adultBMI-", "mvmr_adultBMIearlyBMI-")
mvmr_types <- c("mvmr_sens_BMIs-", "mvmr_sens_earlyBMI-")

# for each mvmr type, load all indiv trait files, merge, save
for (mvmr_type in mvmr_types) {

  print(paste0("Current trait category: ", current_trait_category))
  
  # select type
  mvmr_files_type <- mvmr_files[grepl(mvmr_type, mvmr_files )]
  
  
  print(paste0("Reading data from analysis: ", mvmr_type))
  
  # read all individual mediators
  l <- lapply(mvmr_files_type, read_tsv, show_col_types = FALSE)
  df <- l %>% purrr::reduce(bind_rows) 
  
  file_prefix <- paste0(mvmr_type, current_trait_category, "-to-BCAC")
  write_tsv(df, paste0(results_path, current_trait_category, "/mvmr/merged/merged_sens_", file_prefix, ".tsv"))
  print(paste0(" -> Merged and saved."))

}
```




