---
title: "igf"
author: "Marina Vabistsevits"
date: '2022-07-22'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(data.table)
library(vroom)

library(tidyr)
library(purrr)
library(tibble)
library(dplyr)

library(TwoSampleMR)
library(ggplot2)
library(cowplot)
library(wesanderson)


library(epigraphdb)
#  function that will export query result as a table
query_epigraphdb_as_table <- function(query){
  results_subset <- query_epigraphdb(
    route = "/cypher",
    params = list(query = query),
    method = "POST",
    mode = "table")
}


```

```{r}
instruments<-read_tsv("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project2/01_Data/GWAS_tophits/igf_tophits.tsv")
out <- extract_outcome_data(
  snps = instruments$SNP,
  outcome = 'ieu-a-1126')

harmonised<- harmonise_data(exposure_dat = instruments, 
                            outcome_dat = out)
dim(harmonised)

res <- mr(harmonised, method_list=c('mr_ivw','mr_egger_regression','mr_weighted_median', 'mr_wald_ratio')) %>% 
  split_outcome() %>% 
  split_exposure() %>% 
  generate_odds_ratios()

```

```{r}
p1 <- mr_scatter_plot(res, harmonised)

res_single <- mr_singlesnp(harmonised, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2 <- mr_forest_plot(res_single)


res_loo <- mr_leaveoneout(harmonised)
p3 <- mr_leaveoneout_plot(res_loo)
```

```{r}
library(MRPRESSO)
mr_presso <- mr_presso(data = harmonised,
                       BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", 
                       SdOutcome = "se.outcome", SdExposure = "se.exposure",
                       OUTLIERtest = TRUE, DISTORTIONtest = TRUE,  NbDistribution = 1000,  SignifThreshold = 0.05)
mr_presso_df <- as.data.frame(mr_presso$`Main MR results` )

mr_presso_df <- mr_presso_df %>% rename(b=`Causal Estimate`, se = Sd) %>% generate_odds_ratios()

outlier_indicies <- mr_presso$`MR-PRESSO results`$`Distortion Test`$`Outliers Indices`

#       or or_lci95 or_uci95
# 1.088110 1.029283 1.150298 # raw
# 1.091817 1.044839 1.140907 # outlier corrected


harmonised %>% select(SNP) %>%  mutate(index = row_number()) %>% filter(index %in% outlier_indicies) %>% pull(SNP)
```

```{r}
# testing radial MR -new
library(RadialMR)

# format
raddat <- format_radial(
  harmonised$beta.exposure, harmonised$beta.outcome,
  harmonised$se.exposure, harmonised$se.outcome, 
  harmonised$SNP)

# test IVW 
ivwrad <- ivw_radial(raddat, alpha=0.05/nrow(raddat), weights=3) 
dim(ivwrad$outliers)[1]  #7 outliers at bonf 

ivwrad <- ivw_radial(raddat, alpha=0.05, weights=3)
dim(ivwrad$outliers)[1] #51 outliers at 0.05

# test Egger
eggrad <- egger_radial(raddat, alpha=0.05, weights=3)
eggrad <- egger_radial(raddat, alpha=0.05/nrow(raddat), weights=3)
eggrad$coef 
dim(eggrad$outliers)[1] #52 outliers at 0.05  (7 bonf)

plot_radial(ivwrad,radial_scale = T, show_outliers = F,scale_match= T)
plot_radial(eggrad,radial_scale = T, show_outliers = F,scale_match= T)

plot_radial(c(ivwrad,eggrad), radial_scale = T, show_outliers = F,scale_match= T)


ivwrad$qstatistic 
ivwrad$sortoutliers <- ivwrad$outliers[order(ivwrad$outliers$p.value),]
ivwrad$sortoutliers$Qsum <- cumsum(ivwrad$sortoutliers$Q_statistic)
ivwrad$sortoutliers$Qdif <- ivwrad$sortoutliers$Qsum - ivwrad$qstatistic


#Remove top outliers
harmonised2 <- harmonised[!harmonised$SNP %in% ivwrad$outliers$SNP,]
mr_results2 <- mr(harmonised2) %>%  generate_odds_ratios() # 1.090252 1.0407233 1.142137
```



```{r}
# MR clust
library(mrclust)
browseVignettes("mrclust")
#https://github.com/lindgrengroup/obesity_femrepr_MR/blob/9e39114e2d16bf39adc76040cbb6cff588bbb3dc/clustering/SNP_clustering.R


# Only keep single-SNP results from relevant columns
res_single <- mr_singlesnp(harmonised)
res_single <- res_single %>%
                  select("SNP", "exposure", "outcome",  "b", "se") %>% 
                  filter(grepl("rs", SNP))
dim(res_single)


# Keep relevant columns for clustering
dim(harmonised)
harmonised_sub <- harmonised %>%  select("SNP", "exposure", "outcome",
                                       "beta.exposure", "se.exposure",
                                       "beta.outcome", "se.outcome")
cluster_data <- left_join(harmonised_sub, res_single)



cluster_results <- mr_clust_em(theta = cluster_data$b, 
                                   theta_se = cluster_data$se,
                                   bx = cluster_data$beta.exposure, 
                                   by = cluster_data$beta.outcome,
                                   bxse = cluster_data$se.exposure,
                                   byse = cluster_data$se.outcome,
                                   obs_names = cluster_data$SNP)

cluster_results$results$best
clusters = unique(cluster_results$results$best$cluster_class); # same answer if we type
clusters
head(clusters$cluster_membership$cluster_1)

clust_plot_best = cluster_results$plots$two_stage +
  ggplot2::xlab("Genetic association with IGF1") +
  ggplot2::ylab("Genetic association with BC") +
  ggplot2::ggtitle("")
clust_plot_best



clust_list <- names(cluster_results$cluster_membership)
clust_list <- clust_list[!grepl("Junk",clust_list )]

clust_snp_list <- list()

for (i in clust_list){
  
  x <- names(cluster_results$cluster_membership[[i]])
  y <- x[x!="pr_range_0.1_to_0" ]
  
  a<- cluster_results$cluster_membership[[i]][y]
  snps <- unname(unlist(a))
  
  clust_snp_list[[i]] <- snps
}
  
# do mr by cluster
mr_cluster_list <- list()

for (cluster in names(clust_snp_list)){
  
  yy <- harmonised %>% filter(SNP %in% clust_snp_list[[cluster]])
  yy_mr <- mr(yy, method_list = c("mr_ivw")) %>% 
            generate_odds_ratios() %>%
            select(-lo_ci, -up_ci, id.exposure) %>% 
            mutate(cluster_no = cluster)
  mr_cluster_list[[cluster]] <- yy_mr
  
}

out <- bind_rows(mr_cluster_list)
out

```


```{r}
# map SNPs to genes

mapped_snps_list <- list()

for (i in names(clust_snp_list)){
  
  snp_list <- clust_snp_list[[i]]
  
  query=paste0("
    MATCH (variant:Variant)-[variant_to_gene:VARIANT_TO_GENE]-> (gene:Gene)
    WHERE variant.name in ['", paste0(snp_list, collapse = "', '"),"'] 
    return distinct gene.name, variant.name
    order by gene.name")
  
  results <- query_epigraphdb_as_table(query) %>% mutate(cluster = i)
  
  mapped_snps_list[[i]]<-results
}

mapped_snps_df <- bind_rows(mapped_snps_list)

```

```{r}
enrich_pathways<-function(gene_list, pathway_dbs){

  enriched <- enrichr(gene_list, pathway_dbs)

  for (db_name in names(enriched)){
    if( dim(enriched[[db_name]])[1] > 0){
      enriched[[db_name]]$db <- db_name
    } else  {
      # if it's empty, delete it
      enriched[[db_name]] <- NULL
    }
  }
  enriched_df<-bind_rows(enriched)

  if (dim(enriched_df)[1] > 0){
    enriched_df<- enriched_df %>%
      filter(Adjusted.P.value<0.05) %>% 
      separate(Overlap, into = c("found_genes", "total_genes"), sep="/", remove = F)%>% 
      arrange(Odds.Ratio)
  } else{
    enriched_df<-data.frame()
  }
}

```


```{r}
library(enrichR)
alldbs<-listEnrichrDbs()
pathway_dbs <- c("Reactome_2016", "KEGG_2021_Human", "KEGG_2019_Mouse", "Enrichr_Users_Contributed_Lists_2020", 'DisGeNET' )
#gwa_dbs <- c("GWAS_Catalog_2019", "UK_Biobank_GWAS_v1")
#tissue_db<- c("GTEx_Tissue_Sample_Gene_Expression_Profiles_up","GTEx_Tissue_Sample_Gene_Expression_Profiles_down")

mapped_snps_df %>% count(cluster)

enriched_df<-enrich_pathways(mapped_snps_df %>% filter(cluster == "cluster_1") %>% pull(gene.name), pathway_dbs)

enriched_df_wo_db<-enriched_df %>% dplyr::select(-db) %>% distinct()
plotEnrich(enriched_df_wo_db, showTerms = 40, numChar = 50, y = "Count", orderBy = "P.value")


```

```{r}

mapped_snps_df %>% count(cluster)

clipr::write_clip(mapped_snps_df %>% filter(cluster == "cluster_2") %>% pull(gene.name))

enriched_df<-enrich_pathways(mapped_snps_df %>% filter(cluster == "cluster_2") %>% pull(gene.name), pathway_dbs)

enriched_df_wo_db<-enriched_df %>% dplyr::select(-db) %>% distinct()
plotEnrich(enriched_df_wo_db, showTerms = 40, numChar = 50, y = "Count", orderBy = "P.value")


```


```{r}

mapped_snps_df %>% count(cluster)


enriched_df<-enrich_pathways(mapped_snps_df %>% filter(cluster == "cluster_Null") %>% pull(gene.name), pathway_dbs)

enriched_df_wo_db<-enriched_df %>% dplyr::select(-db) %>% distinct()
plotEnrich(enriched_df_wo_db, showTerms = 40, numChar = 50, y = "Count", orderBy = "P.value")


```

