---
title: "explore instruments"
author: "Marina Vabistsevits"
date: '2022-07-20'
output: html_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(readr)
library(data.table)
library(vroom)

library(tidyr)
library(purrr)
library(tibble)
library(dplyr)

library(TwoSampleMR)
library(ggplot2)
library(cowplot)
library(wesanderson)
```

```{r message=F}
# set path for pre-calculated data, outside the code repo
# `local` / `remote` (reading data from RDSF)
currently_working_env = "local"
source("../set_paths.R")
set_paths(currently_working_env)

# metafile
data_lookup<-read_csv(paste0("../metadata/data_lookup.csv")) 
bcac_lookup<-read_csv(paste0("../metadata/data_lookup_BCAC.csv")) 

# load functions
source("../functions.R")
```


```{r}
da_old <- read_tsv(paste0(data_path_tophits, "mammographic_density_DA_sieh_tophits.tsv"), show_col_types = FALSE) #16
                
da_new <- read_tsv(paste0(data_path_tophits, "dense_area_tophits.tsv")) # 25

da_all <- read_tsv(paste0(data_path_tophits, "dense_area_tophits_all.tsv")) #668

```

```{r}
# DA -new

out <- extract_outcome_data(snps = da_new$SNP, outcome = "ieu-a-1126")
harmonised_new<- harmonise_data(exposure_dat = da_new, outcome_dat = out)
harmonised_new %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()


res_new <- mr(harmonised_new) %>% 
  split_outcome() %>% 
  generate_odds_ratios()

het_new<- mr_heterogeneity(harmonised_new)
egger_int_new <- mr_pleiotropy_test(harmonised_new)


#> Analysing 'ieu-a-2' on 'ieu-a-7'
p1_new <- mr_scatter_plot(res_new, harmonised_new)

res_single_new <- mr_singlesnp(harmonised_new, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2_new <- mr_forest_plot(res_single_new)


res_loo <- mr_leaveoneout(harmonised_new)
p3_new <- mr_leaveoneout_plot(res_loo)

```

```{r}
# DA - old

out <- extract_outcome_data(snps = da_old$SNP, outcome = "ieu-a-1126")
harmonised_old<- harmonise_data(exposure_dat = da_old, outcome_dat = out)

res_old <- mr(harmonised_old) %>% 
  split_outcome() %>% 
  generate_odds_ratios()

het_old<- mr_heterogeneity(harmonised_old)
egger_int_old <- mr_pleiotropy_test(harmonised_old)


res <- mr(harmonised_old)
#> Analysing 'ieu-a-2' on 'ieu-a-7'
p1_old <- mr_scatter_plot(res_old, harmonised_old)

res_single_old <- mr_singlesnp(harmonised_old, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode")) %>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2_old <- mr_forest_plot(res_single_old)


res_loo <- mr_leaveoneout(harmonised_old)
p3_old <- mr_leaveoneout_plot(res_loo)


```

```{r}
# DA -all

out <- extract_outcome_data(snps = da_all$SNP, outcome = "ieu-a-1126")
harmonised_all<- harmonise_data(exposure_dat = da_all, outcome_dat = out)
harmonised_all %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()


res_all <- mr(harmonised_all) %>% 
  split_outcome() %>% 
  generate_odds_ratios()

het_all<- mr_heterogeneity(harmonised_all)
egger_int_all <- mr_pleiotropy_test(harmonised_all)


#> Analysing 'ieu-a-2' on 'ieu-a-7'
p1_all <- mr_scatter_plot(res_all, harmonised_all)

res_single_all <- mr_singlesnp(harmonised_all, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2_all <- mr_forest_plot(res_single_all)


res_loo <- mr_leaveoneout(harmonised_all)
p3_all <- mr_leaveoneout_plot(res_loo)

```



```{r}
library(cowplot)

x<-plot_grid(p1_new[[1]], p2_new[[1]], p3_new[[1]],
             p1_old[[1]], p2_old[[1]], p3_old[[1]], 
             
             labels = c('new', '', "", "old"), nrow=2) 



```


```{r}
# try MR presso
# Load a simulated toy dataset
data(SummaryStats)

# Run MR-PRESSO global method
mrpres <- mr_presso(data = SummaryStats, 
                    BetaOutcome = "Y_effect", BetaExposure = "E1_effect",
                    SdOutcome = "Y_se", SdExposure = "E1_se",
                    OUTLIERtest = TRUE, DISTORTIONtest = TRUE, NbDistribution = 1000,  SignifThreshold = 0.05)

mrpres_res <- as.data.frame(mrpres$`Main MR results` )

mrpres_res%>% rename(b=`Causal Estimate`, se = Sd) %>% generate_odds_ratios()

View(mrpres$`MR-PRESSO results`)

```

```{r}


mr_presso_new <- mr_presso(data = harmonised_new,
                       BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", 
                       SdOutcome = "se.outcome", SdExposure = "se.exposure",
                       OUTLIERtest = TRUE, DISTORTIONtest = TRUE,  NbDistribution = 1000,  SignifThreshold = 0.05)
mr_presso_new_df <- as.data.frame(mr_presso_new$`Main MR results` )

mr_presso_new_df <- mr_presso_new_df %>% rename(b=`Causal Estimate`, se = Sd) %>% generate_odds_ratios()

outlier_indicies_new <- mr_presso_new$`MR-PRESSO results`$`Distortion Test`$`Outliers Indices`


harmonised_new %>% select(SNP) %>%  mutate(index = row_number()) %>% filter(index %in% outlier_indicies_new) %>% pull(SNP)

```

```{r}

mr_presso_old <- mr_presso(data = harmonised_old,
                       BetaOutcome = "beta.outcome", BetaExposure = "beta.exposure", 
                       SdOutcome = "se.outcome", SdExposure = "se.exposure",
                       OUTLIERtest = TRUE, DISTORTIONtest = TRUE,  NbDistribution = 1000,  SignifThreshold = 0.05)

mr_presso_old_df <- as.data.frame(mr_presso_old$`Main MR results` )

mr_presso_old_df <- mr_presso_old_df %>% rename(b=`Causal Estimate`, se = Sd) %>% generate_odds_ratios()


outlier_indicies_old <- mr_presso_old$`MR-PRESSO results`$`Distortion Test`$`Outliers Indices`


harmonised_old %>% select(SNP) %>%  mutate(index = row_number()) %>% filter(index %in% outlier_indicies_old) %>% pull(SNP)

```

```{r}
#F-statistic - new
harmonised_new$samplesize.exposure <- 24192
harmonised_new$samplesize.outcome <- 228951 
harmonised_new <- steiger_filtering(harmonised_new) 

N = harmonised_new$samplesize.exposure[1] #sample size
K = length(harmonised_new$SNP) #number of SNPs
total_r2 <- sum(harmonised_new$rsq.exposure) 
Fstat <- (N-K-1)/K * total_r2 / (1-total_r2)
total_r2 #0.04015256
Fstat #43.95659
```

```{r}
#F-statistic - old
harmonised_old$samplesize.exposure <- 24192
harmonised_old$samplesize.outcome <- 228951 
harmonised_old <- steiger_filtering(harmonised_old) 

N = harmonised_old$samplesize.exposure[1] #sample size
K = length(harmonised_old$SNP) #number of SNPs
total_r2 <- sum(harmonised_old$rsq.exposure) 
Fstat <- (N-K-1)/K * total_r2 / (1-total_r2)
total_r2 #0.02187906
Fstat #36.052
```



```{r}
## estimate r for binary or continouse exposures/outcomes
r_func <- function(x, logistic.exposure,logistic.outcome ){
  # function from: https://github.com/sjfandrews/MR_ADPhenome/blob/b64d8821dbf1546090f47e0642cc8092592cddc8/workflow/scripts/mr_SteigerTest.R#L13
  x$r.exposure <- if(logistic.exposure == TRUE){
    x %>%
      mutate(r.exposure = get_r_from_lor(x$beta.exposure, x$eaf.exposure, x$ncase.exposure, x$ncontrol.exposure, x$prevelance.exposure)) %>% pull(r.exposure)
  } else if(logistic.exposure == FALSE){
    x %>% mutate(r.exposure = get_r_from_pn(x$pval.exposure, x$samplesize.exposure)) %>% pull(r.exposure)
  }

  x$r.outcome <- if(logistic.outcome == TRUE){
    x %>% mutate(r.outcome = get_r_from_lor(x$beta.outcome, x$eaf.outcome, x$ncase.outcome, x$ncontrol.outcome, x$prevelance.outcome)) %>% pull(r.outcome)
  } else if(logistic.outcome == FALSE){
    x %>% mutate(r.outcome = get_r_from_pn(x$pval.outcome, x$samplesize.outcome)) %>% pull(r.outcome)
  }
  x
}



```

```{r}
 ### MR Steiger test of directionality
#https://mrcieu.github.io/TwoSampleMR/articles/perform_mr.html#mr-steiger-directionality-test     ###### VIEW THIS

  bc_out <- extract_outcome_data(snps = da_new$SNP, outcome = "ieu-a-1126")

  # harmonising
  exp_out <- harmonise_data(exposure_dat = da_new, outcome_dat = out)
  exp_out$samplesize.exposure <- 24192
  exp_out$samplesize.outcome <- 228951 
  exp_out$ncase.outcome = 122977
  exp_out$ncontrol.outcome = 105974
  exp_out$prevelance.outcome = 0.53

  exp_out <- r_func(exp_out, logistic.exposure=F, logistic.outcome=T)
  print(nrow(exp_out))


  ## Testing MD on BC
  d_new  <- directionality_test(exp_out)
 #stei<- mr_steiger(p_exp = exp_out$pval.exposure, p_out = exp_out$pval.outcome, 
 #         n_exp = exp_out$samplesize.exposure, n_out = exp_out$samplesize.outcome, 
 #         r_exp = exp_out$beta.exposure, r_out = exp_out$beta.outcome, # these 2 can be 0?
 #         r_xxo = 1, r_yyo = 1)
 #
  bc_out <- extract_outcome_data(snps = da_old$SNP, outcome = "ieu-a-1126")

  # harmonising
  exp_out <- harmonise_data(exposure_dat = da_old, outcome_dat = out)
  exp_out$samplesize.exposure <- 24192
  exp_out$samplesize.outcome <- 228951 
  exp_out$ncase.outcome = 122977
  exp_out$ncontrol.outcome = 105974
  exp_out$prevelance.outcome = 0.53

  exp_out <- r_func(exp_out, logistic.exposure=F, logistic.outcome=T)
  print(nrow(exp_out))
	   
    d_old  <- directionality_test(exp_out)

  
```







```{r}
# testing radial MR -new
library(RadialMR)

# format
raddat_new <- format_radial(
  harmonised_new$beta.exposure, harmonised_new$beta.outcome,
  harmonised_new$se.exposure, harmonised_new$se.outcome, 
  harmonised_new$SNP)

# test IVW 
ivwrad_new <- ivw_radial(raddat_new, alpha=0.05/nrow(raddat_new), weights=3) 
dim(ivwrad_new$outliers)[1]  #9 outliers at bonf 

ivwrad_new <- ivw_radial(raddat_new, alpha=0.05, weights=3)
dim(ivwrad_new$outliers)[1] #13 outliers at 0.05

# test Egger
eggrad_new <- egger_radial(raddat_new, alpha=0.05, weights=3)
eggrad_new <- egger_radial(raddat_new, alpha=0.05/nrow(raddat_new), weights=3)
eggrad_new$coef 
dim(eggrad_new$outliers)[1] #12 outliers at 0.05  (9 bonf)

plot_radial(ivwrad_new,radial_scale = T, show_outliers = F,scale_match= T)
plot_radial(eggrad_new,radial_scale = T, show_outliers = F,scale_match= T)

plot_radial(c(ivwrad_new,eggrad_new), radial_scale = T, show_outliers = F,scale_match= T)


ivwrad_new$qstatistic 
ivwrad_new$sortoutliers <- ivwrad_new$outliers[order(ivwrad_new$outliers$p.value),]
ivwrad_new$sortoutliers$Qsum <- cumsum(ivwrad_new$sortoutliers$Q_statistic)
ivwrad_new$sortoutliers$Qdif <- ivwrad_new$sortoutliers$Qsum - ivwrad_new$qstatistic


#Remove top outliers
harmonised_new2 <- harmonised_new[!harmonised_new$SNP %in% ivwrad_new$outliers$SNP,]
mr_results2 <- mr(harmonised_new2) %>%  generate_odds_ratios()
```


```{r}
# testing radial MR old
library(RadialMR)

# format
raddat_old <- format_radial(
  harmonised_old$beta.exposure, harmonised_old$beta.outcome,
  harmonised_old$se.exposure, harmonised_old$se.outcome, 
  harmonised_old$SNP)

# test IVW 
ivwrad_old <- ivw_radial(raddat_old, alpha=0.05/nrow(raddat_old), weights=3) 
dim(ivwrad_old$outliers)[1]  #4 outliers at bonf 

ivwrad_old <- ivw_radial(raddat_old, alpha=0.05, weights=3)
dim(ivwrad_old$outliers)[1] #7 outliers at 0.05

# test Egger
eggrad_old <- egger_radial(raddat_old, alpha=0.05, weights=3)
eggrad_old <- egger_radial(raddat_old, alpha=0.05/nrow(raddat_old), weights=3)

eggrad_old$coef 
dim(eggrad_old$outliers)[1] #5 outliers at 0.05 

plot_radial(ivwrad_old,radial_scale = T, show_outliers = F,scale_match= T)
plot_radial(c(ivwrad_old,eggrad_old), radial_scale = T, show_outliers = F,scale_match= T)


ivwrad_old$qstatistic 
ivwrad_old$sortoutliers <- ivwrad_old$outliers[order(ivwrad_old$outliers$p.value),]
ivwrad_old$sortoutliers$Qsum <- cumsum(ivwrad_old$sortoutliers$Q_statistic)
ivwrad_old$sortoutliers$Qdif <- ivwrad_old$sortoutliers$Qsum - ivwrad_old$qstatistic


#Remove top outliers
harmonised_old2 <- harmonised_old[!harmonised_old$SNP %in% ivwrad_old$outliers$SNP,]
mr_results2 <- mr(harmonised_old2) %>%  generate_odds_ratios()
```


```{r}
# MR clust
library(mrclust)
#browseVignettes("mrclust")
#https://github.com/lindgrengroup/obesity_femrepr_MR/blob/9e39114e2d16bf39adc76040cbb6cff588bbb3dc/clustering/SNP_clustering.R


# Only keep single-SNP results from relevant columns
res_single_new <- res_single_new %>%
                  select("SNP", "exposure", "outcome",  "b", "se") %>% 
                  filter(grepl("rs", SNP))


# Keep relevant columns for clustering
harmonised_new_sub <- harmonised_new %>%  select("SNP", "exposure", "outcome",
                                       "beta.exposure", "se.exposure",
                                       "beta.outcome", "se.outcome")
cluster_data_new <- left_join(harmonised_new_sub, res_single_new)



cluster_results_new <- mr_clust_em(theta = cluster_data_new$b, 
                                   theta_se = cluster_data_new$se,
                                   bx = cluster_data_new$beta.exposure, 
                                   by = cluster_data_new$beta.outcome,
                                   bxse = cluster_data_new$se.exposure,
                                   byse = cluster_data_new$se.outcome,
                                   obs_names = cluster_data_new$SNP)

cluster_results_new$results$best
clusters_new = unique(cluster_results_new$results$best$cluster_class); # same answer if we type
clusters_new
head(clusters_new$cluster_membership$cluster_1)

clust_plot_best_new = cluster_results_new$plots$two_stage +
  ggplot2::xlab("Genetic association with DA") +
  ggplot2::ylab("Genetic association with BC") +
  ggplot2::ggtitle("")
clust_plot_best_new



clust_list <- names(cluster_results_new$cluster_membership)
clust_list <- clust_list[!grepl("Null|Junk",clust_list )]

clust_snp_list <- list()

for (i in clust_list){
  
  x <- names(cluster_results_new$cluster_membership[[i]])
  y <- x[x!="pr_range_0.1_to_0" ]
  
  a<- cluster_results_new$cluster_membership[[i]][y]
  snps <- unname(unlist(a))
  
  clust_snp_list[[i]] <- snps
}
  
# do mr by cluster
mr_cluster_list <- list()

for (cluster in names(clust_snp_list)){
  
  yy <- harmonised_new %>% filter(SNP %in% clust_snp_list[[cluster]])
  yy_mr <- mr(yy, method_list = c("mr_ivw")) %>% 
            generate_odds_ratios() %>%
            select(-lo_ci, -up_ci, id.exposure) %>% 
            mutate(cluster_no = cluster)
  mr_cluster_list[[cluster]] <- yy_mr
  
}

out_new <- bind_rows(mr_cluster_list)


```

```{r}
# MR-clust old

# Only keep single-SNP results from relevant columns
res_single_old <- res_single_old %>%
                  select("SNP", "exposure", "outcome",  "b", "se") %>% 
                  filter(grepl("rs", SNP))


# Keep relevant columns for clustering
harmonised_old_sub <- harmonised_old %>%  select("SNP", "exposure", "outcome",
                                       "beta.exposure", "se.exposure",
                                       "beta.outcome", "se.outcome")
cluster_data_old <- left_join(harmonised_old_sub, res_single_old)



cluster_results_old <- mr_clust_em(theta = cluster_data_old$b, 
                                   theta_se = cluster_data_old$se,
                                   bx = cluster_data_old$beta.exposure, 
                                   by = cluster_data_old$beta.outcome,
                                   bxse = cluster_data_old$se.exposure,
                                   byse = cluster_data_old$se.outcome,
                                   obs_names = cluster_data_old$SNP)

cluster_results_old$results$best
clusters_old = unique(cluster_results_old$results$best$cluster_class); # same answer if we type
clusters_old
head(cluster_results_old$cluster_membership$cluster_1)

clust_plot_best_old = cluster_results_old$plots$two_stage +
  ggplot2::xlab("Genetic association with DA") +
  ggplot2::ylab("Genetic association with BC") +
  ggplot2::ggtitle("")
clust_plot_best_old


clust_list <- names(cluster_results_old$cluster_membership)
clust_list <- clust_list[!grepl("Null|Junk",clust_list )]

clust_snp_list <- list()

for (i in clust_list){
  
  x <- names(cluster_results_old$cluster_membership[[i]])
  y <- x[x!="pr_range_0.1_to_0" ]
  
  a<- cluster_results_old$cluster_membership[[i]][y]
  snps <- unname(unlist(a))
  
  clust_snp_list[[i]] <- snps
}
  
# do mr by cluster
mr_cluster_list <- list()

for (cluster in names(clust_snp_list)){
  
  yy <- harmonised_old %>% filter(SNP %in% clust_snp_list[[cluster]])
  yy_mr <- mr(yy, method_list = c("mr_ivw")) %>% 
            generate_odds_ratios() %>%
            select(-lo_ci, -up_ci, id.exposure) %>% 
            mutate(cluster_no = cluster)
  mr_cluster_list[[cluster]] <- yy_mr
  
}

out_old <- bind_rows(mr_cluster_list)

```

```{r}
# MR clust on all genome-wide significant SNPs
library(mrclust)

# Only keep single-SNP results from relevant columns
res_single_new <- res_single_all %>%
                  select("SNP", "exposure", "outcome",  "b", "se") %>% 
                  filter(grepl("rs", SNP))


# Keep relevant columns for clustering
harmonised_all_sub <- harmonised_all %>%  select("SNP", "exposure", "outcome",
                                       "beta.exposure", "se.exposure",
                                       "beta.outcome", "se.outcome")
cluster_data_all <- left_join(harmonised_all_sub, res_single_all) %>% drop_na()



cluster_results_all <- mr_clust_em(theta = cluster_data_all$b, 
                                   theta_se = cluster_data_all$se,
                                   bx = cluster_data_all$beta.exposure, 
                                   by = cluster_data_all$beta.outcome,
                                   bxse = cluster_data_all$se.exposure,
                                   byse = cluster_data_all$se.outcome,
                                   obs_names = cluster_data_all$SNP)

cluster_results_all$results$best
clusters_all = unique(cluster_results_all$results$best$cluster_class); # same answer if we type
clusters_all
head(clusters_all$cluster_membership$cluster_1)

clust_plot_best_all = cluster_results_all$plots$two_stage +
  ggplot2::xlab("Genetic association with DA") +
  ggplot2::ylab("Genetic association with BC") +
  ggplot2::ggtitle("")
clust_plot_best_all



clust_list <- names(cluster_results_all$cluster_membership)
clust_list <- clust_list[!grepl("Null|Junk",clust_list )]

clust_snp_list <- list()

for (i in clust_list){
  
  x <- names(cluster_results_all$cluster_membership[[i]])
  y <- x[x!="pr_range_0.1_to_0" ]
  
  a<- cluster_results_all$cluster_membership[[i]][y]
  snps <- unname(unlist(a))
  
  clust_snp_list[[i]] <- snps
}


# list to df
cluster_df <- data.frame(ID = rep(names(clust_snp_list), sapply(clust_snp_list, length)),
                 rsID = unlist(clust_snp_list))
rownames(cluster_df)<-c(1:nrow(cluster_df))
write_tsv(cluster_df, "../../02_Results/tmp/DA_all_snps_clusters.tsv")

  
# do mr by cluster
mr_cluster_list <- list()

for (cluster in names(clust_snp_list)){
  
  yy <- harmonised_all %>% filter(SNP %in% clust_snp_list[[cluster]])
  yy_mr <- mr(yy, method_list = c("mr_ivw")) %>% 
            generate_odds_ratios() %>%
            select(-lo_ci, -up_ci, id.exposure) %>% 
            mutate(cluster_no = cluster)
  mr_cluster_list[[cluster]] <- yy_mr
  
}

out_all <- bind_rows(mr_cluster_list)


```

```{r}
# check FUMA results for all DA snps

genestxt <- read_tsv("../../genes/FUMA_job211823/genes.txt")
snpstxt <- read_tsv("../../genes/FUMA_job211823/snps.txt")
eqtltxt <- read_tsv("../../genes/FUMA_job211823/eqtl.txt")

cluster_df <- read_tsv( "../../02_Results/tmp/DA_all_snps_clusters.tsv")


cluster_df <-cluster_df %>% left_join(snpstxt %>% select(rsID, nearestGene), by = "rsID")

cluster_df_genes<- cluster_df %>% select(cluster=ID, gene = nearestGene) %>% distinct()

cluster_df_genes %>% count(gene,sort=T)


# map to genes with gProfiler online:

grof_res <-  read_csv("../../genes/gProfiler_gSNPense_18-10-2022_11-15-48.csv") %>% select(rsID=id, gprofiler_gene = gene_names) %>% distinct()

cluster_df<- left_join(cluster_df,grof_res, by ="rsID")

cluster_df_genes2<- cluster_df %>% select(cluster=ID, geneFUMA = nearestGene, gprofiler_gene) %>% distinct()

write_tsv(cluster_df_genes2, "../../02_Results/tmp/DA_all_snps_clusters_gene_annotation.tsv")


### Querying eQTL catalogue for  variats

source("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/PhD_project/code/epigraphdb-breast-cancer/project_code/helper_functions.R")

eqtl1 <- eqtl_for_snps(unique(cluster_df$rsID))


eqtl1_sub <- eqtl1 %>% select(rsid, hgnc_symbol) %>% distinct()

length(unique(cluster_df$rsID)) #563
length(unique(eqtl1_sub$rsid)) # 375

eqtl1_sub %>% count(rsid, sort=T) %>% View()

eqtl1 %>% count(qtl_group, sort=T) %>% View()

tissues_keep <-c(
"Cells_Cultured_fibroblasts","fibroblast",
"adipose_naive" ,"Adipose_Subcutaneous" ,"Adipose_Visceral_Omentum" ,"fat",
"blood","Whole_Blood","Breast_Mammary_Tissue" )

eqtl1_sub_selected_tissues <- eqtl1 %>% filter(qtl_group %in% tissues_keep) %>% 
                                        select(rsid, hgnc_symbol,qtl_group, median_tpm, pvalue, study_id) %>% distinct() %>% 
                                        filter(median_tpm > 0) 
length(unique(eqtl1_sub_selected_tissues$rsid)) # 98


cluster_df<- left_join(cluster_df,eqtl1_sub_selected_tissues, by =c("rsID"="rsid"))

cluster_df_genes3<- cluster_df %>% select(cluster=ID, geneFUMA = nearestGene, gprofiler_gene, eQTL_gene = hgnc_symbol, everything()) %>% distinct() %>% mutate(eQTL_gene = ifelse(eQTL_gene=="", NA, eQTL_gene))

cluster_df_genes3_sub<- cluster_df_genes3 %>% select("cluster" , "geneFUMA","gprofiler_gene" ,"eQTL_gene" ) %>% distinct()

write_tsv(cluster_df_genes3_sub, "../../02_Results/tmp/DA_all_snps_clusters_gene_annotation_plus_eQTL.tsv")

```


```{r}
#phenoscanner
library(phenoscanner)
library(ieugwasr)
library(TwoSampleMR)


get_pheno_assoc <- function(snp){

  # query phenoscanner
  res <- phenoscanner::phenoscanner(snpquery=snp)
  phenosc <- res$results %>% filter(ancestry == "European")
  
  if (nrow(phenosc) > 0){
    phenosc<- phenosc %>%  as.data.frame() %>% 
      filter(p <= 5e-8) %>% arrange(p) %>% pull(trait) %>% unique()
  } else {  phenosc<-c() } # in case no res
  
  # query opengwas
  ao <-TwoSampleMR::available_outcomes() %>% filter(population == "European")
  phwnogw<- ieugwasr::phewas(variants=snp, pval=5e-8) %>% filter(id %in% ao$id)
  if (nrow(phwnogw) > 0){
      phwnogw <- phwnogw %>%  as.data.frame()  %>% 
        arrange(p) %>% pull(trait) %>% unique()
  } else {  phwnogw<-c() }  # in case no res
    
  # merge sources
  out <- unique(c(phenosc, phwnogw))
    
}

xx <- get_pheno_assoc(snp="rs11877925", ao)

# snps: 
snp_df<- unlist(clust_snp_list) %>% as_data_frame()


snp_phewas <- lapply(snp_df$value, get_pheno_assoc )
names(snp_phewas) <- snp_df$value


save(snp_phewas, file="../02_Results/cluster_phewas.RData")

View(snp_phewas[["rs11205303"]])

```












```{r, message=F}
library(enrichR)
alldbs<-listEnrichrDbs()
pathway_dbs <- c("Reactome_2016", "KEGG_2021_Human", 
                 "GO_Biological_Process_2021", "GO_Cellular_Component_2021", "GO_Molecular_Function_2021",
                  "WikiPathway_2021_Human")

## function to call enrich fro pathways
enrich_dbs<-function(gene_list, dbs, adjpval_filter = 0.05){
  
  enriched <- enrichr(gene_list, dbs)
  # flatten list into a table; handle empty tables
  for (db_name in names(enriched)){
    if( dim(enriched[[db_name]])[1] > 0){
      enriched[[db_name]]$db <- db_name
    } else  {
      # if it's empty, delete it
      enriched[[db_name]] <- NULL
    }
  }
  enriched_df<-bind_rows(enriched)
  if (dim(enriched_df)[1] > 0){
    enriched_df<- enriched_df %>%
      filter(Adjusted.P.value < adjpval_filter) %>% 
      separate(Overlap, into = c("found_genes", "total_genes"), sep="/", remove = F)%>% 
      arrange(Odds.Ratio)
  } else{
    enriched_df<-data.frame()
  }
}
kable_it<-function(df){
  library(kableExtra)
  df %>% 
    kable(.) %>%
    kable_paper("hover", full_width = F)
}


```


```{r}
# load SNP to gene mapping (collected from different sources)
DA_snps_to_genes <- read_csv("../../genes/DA_snp_to_gene.csv")
cluster1_genes <- DA_snps_to_genes %>% filter(CLUSTER==1) %>% pull(final)
cluster2_genes <- DA_snps_to_genes %>% filter(CLUSTER==2) %>% pull(final)
cluster3_genes <- DA_snps_to_genes %>% filter(CLUSTER==3) %>% pull(final)
```


```{r}
# testing enrichr on DA snps

gene_list <- DA_snps_to_genes %>% select(final) %>% drop_na() %>% pull(final)
enriched <- enrichr(gene_list, pathway_dbs)
View(enriched$KEGG_2021_Human)


  # flatten list into a table; handle empty tables
  for (db_name in names(enriched)){
    if( dim(enriched[[db_name]])[1] > 0){
      enriched[[db_name]]$db <- db_name
    } else  {
      # if it's empty, delete it
      enriched[[db_name]] <- NULL
    }
  }
  enriched_df<-bind_rows(enriched) %>% select( Term, Genes, db) 
  

enriched_df_unw<- enriched_df %>% tidyr::separate_rows(Genes, sep = ";", convert = TRUE)


enriched_df_unw %>%  filter(grepl("MTMR11|OSR1|RALB|MRTFA", Genes)) %>% View()

library(eulerr)

# negative and junk
s1 <- list(MTMR11 = enriched_df_unw %>% filter(Genes == "MTMR11") %>% pull(Term) %>% sort(),
           OSR1 = enriched_df_unw %>% filter(Genes == "OSR1") %>% pull(Term) %>% sort(),
           RALB = enriched_df_unw %>% filter(Genes == "RALB") %>% pull(Term) %>% sort(),
           MRTFA = enriched_df_unw %>% filter(Genes == "MRTFA") %>% pull(Term) %>% sort())
plot(euler(s1, shape = "ellipse"), quantities = TRUE)

s1_list<-unique(unlist(s1))

s2 <- list(RPL17  = enriched_df_unw %>% filter(Genes == "RPL17") %>% pull(Term) %>% sort(),
           BCL2  = enriched_df_unw %>% filter(Genes  == "BCL2") %>% pull(Term) %>% sort(),
           AREG  = enriched_df_unw %>% filter(Genes  == "AREG") %>% pull(Term) %>% sort(),
           SMIM25  = enriched_df_unw %>% filter(Genes== "SMIM25") %>% pull(Term) %>% sort(),
           PRKG1  = enriched_df_unw %>% filter(Genes == "PRKG1") %>% pull(Term) %>% sort(),
           MKX  = enriched_df_unw %>% filter(Genes   == "MKX") %>% pull(Term) %>% sort(),
           PRDM6  = enriched_df_unw %>% filter(Genes == "PRDM6") %>% pull(Term) %>% sort(),
           ATXN1  = enriched_df_unw %>% filter(Genes == "ATXN1") %>% pull(Term) %>% sort(),
           TAB2  = enriched_df_unw %>% filter(Genes  == "TAB2") %>% pull(Term) %>% sort(),
           FAF1  = enriched_df_unw %>% filter(Genes  == "FAF1") %>% pull(Term) %>% sort(),
           TET3  = enriched_df_unw %>% filter(Genes  == "TET3") %>% pull(Term) %>% sort(),
           PPARG  = enriched_df_unw %>% filter(Genes == "PPARG") %>% pull(Term) %>% sort(),
           TEX43  = enriched_df_unw %>% filter(Genes == "TEX43") %>% pull(Term) %>% sort(),
           FUT2  = enriched_df_unw %>% filter(Genes  == "FUT2") %>% pull(Term) %>% sort())
plot(euler(s2, shape = "ellipse"), quantities = TRUE)

s2_list<-unique(unlist(s2))



# extra positive
s4 <- list(ZNF365 = enriched_df_unw %>% filter(Genes == "ZNF365") %>% pull(Term) %>% sort(),
           DIRC3 = enriched_df_unw %>% filter(Genes == "DIRC3") %>% pull(Term) %>% sort(),
           RAD51B = enriched_df_unw %>% filter(Genes == "RAD51B") %>% pull(Term) %>% sort())
plot(euler(s4, shape = "circle"), quantities = TRUE)
s4_list<-unique(unlist(s4))

# both

s5 <- list(negative = s1_list,
           positive = s2_list,
           positive_extra = s4_list)
plot(euler(s5, shape = "circle"), quantities = TRUE)


only_in_neg <- setdiff(s1_list, c(s2_list, s4_list))
length(only_in_neg) #109

enriched_df %>% filter(Term %in% only_in_neg) %>% View()
enriched_df %>% filter(Term %in% only_in_neg) %>% count(Term) %>% View()

enriched_df %>% filter(!Term %in% only_in_neg) %>% View()
```

```{r}
# testing enrichr on DA snps --- all


cluster_df_genes2<- read_tsv("../../02_Results/tmp/DA_all_snps_clusters_gene_annotation_upd.tsv") 
cluster_df_genes2_sub <- cluster_df_genes2 %>% select(cluster, gene) %>% drop_na() %>% distinct()


gene_list <- cluster_df_genes2_sub %>% pull(gene) %>% unique()
enriched <- enrichr(gene_list, pathway_dbs)
View(enriched$KEGG_2021_Human)


# flatten list into a table; handle empty tables
for (db_name in names(enriched)){
  if( dim(enriched[[db_name]])[1] > 0){
    enriched[[db_name]]$db <- db_name
  } else  {
    # if it's empty, delete it
    enriched[[db_name]] <- NULL
  }
}
enriched_df<-bind_rows(enriched) %>% select( Term, Genes, db) 

enriched_df_unw<- enriched_df %>% tidyr::separate_rows(Genes, sep = ";", convert = TRUE)

genes_not_mapped <- setdiff(gene_list, unique(enriched_df_unw$Genes)) # need to eQTL those genes
# "GVINP1"    "PELATON"   "LINC01270" "HSPE1P10"  "FNBP1P1"   "DIRC3"     "LINC01376"


enriched_df_unw <-enriched_df_unw %>% left_join(cluster_df_genes2_sub, by  = c("Genes" = "gene"))

enriched_df_unw <-enriched_df_unw %>% 
                        mutate(clust_type = case_when(grepl("6|3|4", cluster) ~ "negative",
                                                      grepl("1|5|2", cluster) ~ "positive")) 

enriched_df_unw_clust <-enriched_df_unw %>%  select(-Genes, -cluster) %>% distinct()

library(eulerr)
# negative and junk
sX <- list(positive = enriched_df_unw_clust %>% filter(clust_type == "positive") %>% pull(Term) %>% sort(),
           negative = enriched_df_unw_clust %>% filter(clust_type == "negative") %>% pull(Term) %>% sort())
plot(euler(sX, shape = "ellipse"), quantities = TRUE)

length(setdiff(sX$negative, sX$positive))
onlyneg<- setdiff(sX$negative, sX$positive)# 87 # 141 with OSR1 #138 other manually added
onlypos<- setdiff(sX$positive, sX$negative)#600 # 584 #590
bothclust<- intersect(sX$positive, sX$negative)#31 # 47 # 126

enriched_df_unw %>% filter(Term %in% onlyneg) %>% View()


path_count<- enriched_df_unw %>% count(Term, sort=T)

enriched_df_unw <- enriched_df_unw %>%
  mutate(clust_status =
         case_when(Term %in% onlyneg ~ "only in negative",
                   Term %in% onlypos ~ "only in positive",
                   TRUE ~ "in both ")) %>% 
  left_join(path_count, by="Term")


library(eulerr)

# negative and junk
s1 <- list(MTMR11 = enriched_df_unw %>% filter(Genes == "MTMR11") %>% pull(Term) %>% sort(),
           OSR1 = enriched_df_unw %>% filter(Genes == "OSR1") %>% pull(Term) %>% sort(),
           RALB = enriched_df_unw %>% filter(Genes == "RALB") %>% pull(Term) %>% sort(),
           MRTFA = enriched_df_unw %>% filter(Genes == "MRTFA") %>% pull(Term) %>% sort())
plot(euler(s1, shape = "ellipse"), quantities = TRUE)

s1_list<-unique(unlist(s1))

s2 <- list(RPL17  = enriched_df_unw %>% filter(Genes == "RPL17") %>% pull(Term) %>% sort(),
           BCL2  = enriched_df_unw %>% filter(Genes  == "BCL2") %>% pull(Term) %>% sort(),
           AREG  = enriched_df_unw %>% filter(Genes  == "AREG") %>% pull(Term) %>% sort(),
           SMIM25  = enriched_df_unw %>% filter(Genes== "SMIM25") %>% pull(Term) %>% sort(),
           PRKG1  = enriched_df_unw %>% filter(Genes == "PRKG1") %>% pull(Term) %>% sort(),
           MKX  = enriched_df_unw %>% filter(Genes   == "MKX") %>% pull(Term) %>% sort(),
           PRDM6  = enriched_df_unw %>% filter(Genes == "PRDM6") %>% pull(Term) %>% sort(),
           ATXN1  = enriched_df_unw %>% filter(Genes == "ATXN1") %>% pull(Term) %>% sort(),
           TAB2  = enriched_df_unw %>% filter(Genes  == "TAB2") %>% pull(Term) %>% sort(),
           FAF1  = enriched_df_unw %>% filter(Genes  == "FAF1") %>% pull(Term) %>% sort(),
           TET3  = enriched_df_unw %>% filter(Genes  == "TET3") %>% pull(Term) %>% sort(),
           PPARG  = enriched_df_unw %>% filter(Genes == "PPARG") %>% pull(Term) %>% sort(),
           TEX43  = enriched_df_unw %>% filter(Genes == "TEX43") %>% pull(Term) %>% sort(),
           FUT2  = enriched_df_unw %>% filter(Genes  == "FUT2") %>% pull(Term) %>% sort())
plot(euler(s2, shape = "ellipse"), quantities = TRUE)

s2_list<-unique(unlist(s2))



# extra positive
s4 <- list(ZNF365 = enriched_df_unw %>% filter(Genes == "ZNF365") %>% pull(Term) %>% sort(),
           DIRC3 = enriched_df_unw %>% filter(Genes == "DIRC3") %>% pull(Term) %>% sort(),
           RAD51B = enriched_df_unw %>% filter(Genes == "RAD51B") %>% pull(Term) %>% sort())
plot(euler(s4, shape = "circle"), quantities = TRUE)
s4_list<-unique(unlist(s4))

# both

s5 <- list(negative = s1_list,
           positive = s2_list,
           positive_extra = s4_list)
plot(euler(s5, shape = "circle"), quantities = TRUE)


only_in_neg <- setdiff(s1_list, c(s2_list, s4_list))
length(only_in_neg) #109

enriched_df %>% filter(Term %in% only_in_neg) %>% View()
enriched_df %>% filter(Term %in% only_in_neg) %>% count(Term) %>% View()

enriched_df %>% filter(!Term %in% only_in_neg) %>% View()
```


<br>

### Pathways enrichment analysis

```{r}
enriched_df<-enrich_dbs(cluster3_genes, pathway_dbs)
enriched_df_wo_db<-enriched_df %>% select(-db) %>% distinct()
plotEnrich(enriched_df_wo_db, showTerms = 20, numChar = 50, y = "Count", orderBy = "P.value")
```

Table sorted by odds ratio:
```{r , echo=F}
numChar=30
enriched_df %>% 
  select(Term, Overlap, 'P.value','Adjusted.P.value', 'Odds.Ratio', Genes, db) %>% 
  arrange(desc(Odds.Ratio)) %>% 
  mutate(Term = paste(substr(Term, 1, numChar), ifelse(nchar(Term) > numChar, "...", ""), sep = ""),
         Odds.Ratio = round(Odds.Ratio, digits=1), 
         P.value= scales::scientific(P.value, digits = 2),
         Adjusted.P.value= scales::scientific(Adjusted.P.value, digits = 2)) %>% 
  rename(Adj.P.value= Adjusted.P.value) %>% 
  kable_it()
```


#### Enrichment GWAS-eQTL genes in pathways, using g:Profiler
```{r}
library(gprofiler2)
gostres <- gost(query = gene_list,
                organism = "hsapiens",
                significant = F)

p <- gostplot(gostres, capped = FALSE, interactive = T)
p
```

g:Profiler results as a table:
```{r}
# recall = intersection_size/term_size
gostres$result %>% arrange(-recall) %>% 
  select("source", "term_name",p_value,  "term_size", "intersection_size", recall) %>% 
  mutate(recall=round(recall,2),
         p_value= scales::scientific(p_value, digits = 2)) %>% 
  kable_it()
```

# doing a multi-query, with named list, 1 gene per query, to get all apths linked to each gene

```{r}
g_list<-as.list(gene_list)
names(g_list) <- gene_list

gostres_multi <- gost(query = g_list,
                organism = "hsapiens",
                significant = F, multi_query=F)

```






```{r}
# check association with BMI

# UKB
adult_bmi_ukb_gwas <- vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project4/01_Data/GWAS_tidy/adult_bmi_cont_ukb_GWAS_tidy_outcome.txt.gz")
# GIANT
out <- extract_outcome_data(snps = da_new$SNP, outcome = "ieu-a-974")


harmonised_bmi<- harmonise_data(exposure_dat = da_new, outcome_dat = adult_bmi_ukb_gwas)


harmonised_bmi %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()


res_bmi <- mr(harmonised_bmi) %>% 
  split_outcome() %>% 
  generate_odds_ratios()

het_bmi<- mr_heterogeneity(harmonised_bmi)
egger_int_bmi <- mr_pleiotropy_test(harmonised_bmi)


#> Analysing 'ieu-a-2' on 'ieu-a-7'
p1_bmi <- mr_scatter_plot(res_bmi, harmonised_bmi)

res_single_bmi <- mr_singlesnp(harmonised_bmi, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2_bmi <- mr_forest_plot(res_single_bmi)


res_loo <- mr_leaveoneout(harmonised_bmi)
p3_bmi <- mr_leaveoneout_plot(res_loo)


# check fat/fat free mass

# UKB
fatmass_gwas <- vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project4/01_Data/GWAS_tidy/fat_mass_GWAS_tidy_outcome.txt.gz")

harmonised_fatmass<- harmonise_data(exposure_dat = da_new, outcome_dat = fatmass_gwas)
harmonised_fatmass %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()

res_single <- mr_singlesnp(harmonised_fatmass, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2fat <- mr_forest_plot(res_single)


fatfreemass_gwas <- vroom("/Users/ny19205/OneDrive - University of Bristol/Documents - OneDrive/Mini-project4/01_Data/GWAS_tidy/fatfree_mass_GWAS_tidy_outcome.txt.gz")

harmonised_fatfreemass<- harmonise_data(exposure_dat = da_new, outcome_dat = fatfreemass_gwas)
harmonised_fatfreemass %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()
res_single <- mr_singlesnp(harmonised_fatfreemass, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2fatfree <- mr_forest_plot(res_single)




```

```{r}
#MR without SNPs associted with BC

out <- extract_outcome_data(snps = da_new$SNP, outcome = "ieu-a-1126") %>% filter(pval.outcome>5e-08) # remove SNPs that are associted with BC

da_new_wo_bc <- da_new %>% filter(SNP %in% out$SNP) # only keeping SNPs not associated with BC


out <- extract_outcome_data(snps = da_new_wo_bc$SNP, outcome = "ieu-a-1126")
harmonised_new<- harmonise_data(exposure_dat = da_new_wo_bc, outcome_dat = out)
harmonised_new %>% select(SNP, beta.exposure, beta.outcome, pval.exposure, pval.outcome) %>% View()


res_new <- mr(harmonised_new) %>% 
  split_outcome() %>% 
  generate_odds_ratios() # 1.385758 1.2366010 1.552906

het_new<- mr_heterogeneity(harmonised_new) #47.66042   14 1.495993e-05
egger_int_new <- mr_pleiotropy_test(harmonised_new) #  -0.0009239941 



#> Analysing 'ieu-a-2' on 'ieu-a-7'
p1_new <- mr_scatter_plot(res_new, harmonised_new)

res_single_new <- mr_singlesnp(harmonised_new, all_method=c("mr_ivw_fe","mr_ivw_mre","mr_egger_regression","mr_weighted_median","mr_weighted_mode"))%>% 
  mutate(SNP = gsub("Inverse variance weighted", "IVW", SNP))
p2_new <- mr_forest_plot(res_single_new)


res_loo <- mr_leaveoneout(harmonised_new)
p3_new <- mr_leaveoneout_plot(res_loo)

```

